// @microsoft/sp-build-core-tasks/lib/index.js
"use strict";
var ConfigureWebpackTask_1 = require("./configureWebpack/ConfigureWebpackTask");
var ConfigureExternalBundlingWebpackTask_1 = require("./configureWebpack/ConfigureExternalBundlingWebpackTask");
var CopyAssetsTask_1 = require("./copyAssets/CopyAssetsTask");
var WriteManifestsTask_1 = require("./writeManifests/WriteManifestsTask");
var CasperJSTask_1 = require("./casperJs/CasperJSTask");
var CopyStaticAssetsTask_1 = require("./copyStaticAssets/CopyStaticAssetsTask");
var CollectLocalizedResourcesTask_1 = require("./collectLocalizedResources/CollectLocalizedResourcesTask");
var PackageSolutionTask_1 = require("./packageSolution/PackageSolutionTask");
var DeployAzureStorageTask_1 = require("./deployAzureStorage/DeployAzureStorageTask");
var DevDeployTask_1 = require("./devDeploy/DevDeployTask");
exports.configureWebpack = new ConfigureWebpackTask_1.ConfigureWebpackTask();
exports.configureExternalBundlingWebpack = new ConfigureExternalBundlingWebpackTask_1.ConfigureExternalBundlingWebpackTask();
exports.copyAssets = new CopyAssetsTask_1.CopyAssetsTask();
exports.writeManifests = new WriteManifestsTask_1.WriteManifestsTask();
exports.casperJSRunner = new CasperJSTask_1.CasperJSTask();
exports.copyStaticAssets = new CopyStaticAssetsTask_1.CopyStaticAssetsTask();
exports.collectLocalizedResources = new CollectLocalizedResourcesTask_1.CollectLocalizedResourcesTask();
exports.packageSolution = new PackageSolutionTask_1.PackageSolutionTask();
exports.deployAzureStorage = new DeployAzureStorageTask_1.DeployAzureStorageTask();
exports.devDeploy = new DevDeployTask_1.DevDeployTask();
