// @microsoft/api-extractor/lib/ast/AstParameter.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
const AstItem_1 = require("./AstItem");
/**
 * This class is part of the AstItem abstract syntax tree. It represents parameters of a function declaration
 */
class AstParameter extends AstItem_1.default {
    constructor(options, docComment) {
        super(options);
        this.kind = AstItem_1.AstItemKind.Parameter;
        const parameterDeclaration = options.declaration;
        this.isOptional = !!parameterDeclaration.questionToken || !!parameterDeclaration.initializer;
        if (parameterDeclaration.type) {
            this.type = parameterDeclaration.type.getText();
        }
        else {
            this.hasIncompleteTypes = true;
            this.type = 'any';
        }
        this.isSpread = !!parameterDeclaration.dotDotDotToken;
    }
}
exports.default = AstParameter;
