// @microsoft/sp-webpart-base/lib/propertyPane/propertyPane/PropertyPaneDefinitions.js
"use strict";
var WebPartConfigurationEvent;
(function (WebPartConfigurationEvent) {
    WebPartConfigurationEvent[WebPartConfigurationEvent["PropertyPaneConfigurationStart"] = 1] = "PropertyPaneConfigurationStart";
    WebPartConfigurationEvent[WebPartConfigurationEvent["PropertyPaneConfigurationComplete"] = 2] = "PropertyPaneConfigurationComplete";
    WebPartConfigurationEvent[WebPartConfigurationEvent["PropertyPaneOpened"] = 3] = "PropertyPaneOpened";
    WebPartConfigurationEvent[WebPartConfigurationEvent["PropertyPaneClosed"] = 4] = "PropertyPaneClosed";
    WebPartConfigurationEvent[WebPartConfigurationEvent["PropertyPaneApplyClicked"] = 5] = "PropertyPaneApplyClicked";
    WebPartConfigurationEvent[WebPartConfigurationEvent["PropertyPaneLostFocus"] = 6] = "PropertyPaneLostFocus";
    WebPartConfigurationEvent[WebPartConfigurationEvent["PropertyPaneWebPartsSwitched"] = 7] = "PropertyPaneWebPartsSwitched";
})(WebPartConfigurationEvent = exports.WebPartConfigurationEvent || (exports.WebPartConfigurationEvent = {}));
var PropertyPaneState;
(function (PropertyPaneState) {
    PropertyPaneState[PropertyPaneState["Open"] = 1] = "Open";
    PropertyPaneState[PropertyPaneState["Close"] = 2] = "Close";
    PropertyPaneState[PropertyPaneState["Toggle"] = 3] = "Toggle";
    PropertyPaneState[PropertyPaneState["Default"] = 4] = "Default";
    PropertyPaneState[PropertyPaneState["Refresh"] = 5] = "Refresh";
})(PropertyPaneState = exports.PropertyPaneState || (exports.PropertyPaneState = {}));
