// @microsoft/sp-module-interfaces/lib/manifestSchemaValidator.js
"use strict";
exports.ZSchema = require("z-schema");
// @todo 264263 - utilize the schema validator in sp-build-common
var fs = require("fs");
var os_1 = require("os");
var path = require("path");
// Discover the schema files
var schemasDirectory = path.join(__dirname, 'manifestSchemas', 'jsonSchemas');
var schemaFiles = fs.readdirSync(schemasDirectory);
var schemas = schemaFiles.map(function (schemaFile) { return require(path.join(schemasDirectory, schemaFile)); });
var manifestSchema = require('./manifestSchemas/jsonSchemas/clientSideComponentManifestSchema.json');
var applicationSchema = require('./manifestSchemas/jsonSchemas/clientSideApplicationManifestSchema.json');
var webPartSchema = require('./manifestSchemas/jsonSchemas/clientSideWebPartManifestSchema.json');
var codePartSchema = require('./manifestSchemas/jsonSchemas/clientSideCodePartManifestSchema.json');
var librarySchema = require('./manifestSchemas/jsonSchemas/clientSideLibraryManifestSchema.json');
exports.zSchemaOptions = {
    breakOnFirstError: true,
    forceAdditional: true,
    forceItems: true,
    forceMaxLength: false,
    forceProperties: false,
    noExtraKeywords: true,
    noTypeless: true,
    noEmptyStrings: true
};
var schemaValidator = new exports.ZSchema(exports.zSchemaOptions);
// @todo 272561 - update the z-schema typings
schemaValidator.validateSchema(schemas); // tslint:disable-line:no-any
var ManifestValidator = (function () {
    function ManifestValidator() {
    }
    ManifestValidator.validateApplicationManifest = function (manifest) {
        return ManifestValidator._validateManifest(manifest, applicationSchema);
    };
    ManifestValidator.validateCodePartManifest = function (manifest) {
        return ManifestValidator._validateManifest(manifest, codePartSchema);
    };
    /* tslint:disable:no-any */
    ManifestValidator.validateWebPartManifest = function (manifest) {
        return ManifestValidator._validateManifest(manifest, webPartSchema);
    };
    ManifestValidator.validateLibraryManifest = function (manifest) {
        return ManifestValidator._validateManifest(manifest, librarySchema);
    };
    ManifestValidator.validateManifest = function (manifest) {
        var parsedManifest = (typeof manifest === 'string')
            ? JSON.parse(manifest)
            : manifest;
        return ManifestValidator._validateManifest(parsedManifest, manifestSchema);
    };
    ManifestValidator.extractInnerErrorMessages = function (errors) {
        var printZSchemaError = function (error) {
            /* tslint:disable:no-any */
            var innerErrors = [];
            (error.inner || []).forEach(function (innerErr) {
                innerErrors = innerErrors.concat(printZSchemaError(innerErr));
            });
            /* tslint:enable:no-any */
            return ["(" + error.path + ") " + error.message].concat(innerErrors);
        };
        var errorList = [];
        errors.map(function (error) { errorList = errorList.concat(printZSchemaError(error)); });
        return errorList;
    };
    ManifestValidator.getFormattedErrorMessage = function (errors) {
        return this.extractInnerErrorMessages(errors).join(os_1.EOL);
    };
    ManifestValidator._validateManifest = function (manifest, schema) {
        if (!schema) {
            throw new Error('Unable to find the manifest schema.');
        }
        if (typeof manifest === 'string') {
            manifest = JSON.parse(manifest);
        }
        var result = schemaValidator.validate(manifest, schema);
        return {
            result: result,
            errors: schemaValidator.getLastErrors() || []
        };
    };
    return ManifestValidator;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = ManifestValidator;
