// @microsoft/sp-build-core-tasks/lib/packageSolution/packageSolution/utils/readSharepointAssets.js
/**
 * @file readSharepointAssets.ts
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 *
 * Using the configuration, reads a list of Sharepoint assets of one of 3 types
 * (element manifests, element files, and upgrade actions). Returns an object with a list
 * of IXml files to include in the package.
 *
 */
"use strict";
var path = require("path");
var fs = require("fs");
var colors = require("colors");
var gulp_core_build_1 = require("@microsoft/gulp-core-build");
/**
 * Reads a special folder containing SharePoint assets which should automatically be integrated into a feature.
 * You can think of this as a slightly more useful version of readCustomFeatures, which allows for custom features
 * to be created, but is more cumbersome when you want to do something simple like include an upgrade action into
 * the SPPKG.
 */
// tslint:disable-next-line:max-line-length
function readSharepointAssets(assets, featureId) {
    return {
        elementManifests: readAssetFiles(assets.elementManifests, 'Element Manifest', featureId),
        elementFiles: readAssetFiles(assets.elementFiles, 'Element File'),
        upgradeActions: readAssetFiles(assets.upgradeActions, 'Upgrade Action', featureId)
    };
}
exports.readSharepointAssets = readSharepointAssets;
function readAssetFiles(assetFilePaths, assetFileDescription, subdir) {
    return assetFilePaths.map(function (filepath) {
        var content;
        var filename = path.basename(filepath);
        try {
            content = fs.readFileSync(filepath).toString();
        }
        catch (err) {
            throw new Error("Could not find " + assetFileDescription + ": " + filepath);
        }
        gulp_core_build_1.log(colors.cyan("Found " + assetFileDescription + ": " + path.basename(filename)));
        return {
            filename: subdir ? path.posix.join(subdir, filename) : filename,
            xml: content
        };
    });
}
