// @microsoft/sp-page-context/lib/pageManager/SharePointPageContextDataProvider.js
"use strict";
var sp_http_1 = require("@microsoft/sp-http");
var SharePointPageContextDataProvider = (function () {
    function SharePointPageContextDataProvider(serviceScope) {
        var _this = this;
        serviceScope.whenFinished(function () {
            _this._httpClient = serviceScope.consume(sp_http_1.SPHttpClient.serviceKey);
        });
    }
    SharePointPageContextDataProvider._createPageContextData = function (spPageContextInfo) {
        return {
            cultureInfo: {
                currentCultureName: spPageContextInfo.currentCultureName,
                currentUICultureName: spPageContextInfo.currentUICultureName,
                isRightToLeft: document.documentElement.dir === 'rtl' ?
                    true :
                    false
            },
            list: spPageContextInfo.listId && spPageContextInfo.listId !== '' ? {
                baseTemplate: spPageContextInfo.listBaseTemplate,
                id: spPageContextInfo.listId,
                permissions: spPageContextInfo.listPermsMask,
                serverRelativeUrl: spPageContextInfo.listUrl,
                title: spPageContextInfo.listTitle
            } : undefined,
            listItem: spPageContextInfo.pageItemId && spPageContextInfo.pageItemId !== -1 ? {
                id: spPageContextInfo.pageItemId
            } : undefined,
            site: {
                absoluteUrl: spPageContextInfo.siteAbsoluteUrl,
                cdnPrefix: spPageContextInfo.cdnPrefix,
                classification: spPageContextInfo.siteClassification,
                correlationId: spPageContextInfo.CorrelationId,
                id: spPageContextInfo.siteId,
                isNoScriptEnabled: spPageContextInfo.isNoScriptEnabled,
                recycleBinItemCount: spPageContextInfo.RecycleBinItemCount,
                serverRelativeUrl: spPageContextInfo.siteServerRelativeUrl,
                serverRequestPath: spPageContextInfo.serverRequestPath,
                sitePagesEnabled: spPageContextInfo.sitePagesEnabled
            },
            user: {
                isAnonymousGuestUser: spPageContextInfo.isAnonymousGuestUser,
                isExternalGuestUser: spPageContextInfo.isExternalGuestUser,
                displayName: spPageContextInfo.userDisplayName,
                email: spPageContextInfo.userEmail,
                loginName: spPageContextInfo.userLoginName
            },
            web: {
                absoluteUrl: spPageContextInfo.webAbsoluteUrl,
                id: spPageContextInfo.webId,
                isAppWeb: spPageContextInfo.isAppWeb,
                language: spPageContextInfo.webLanguage,
                logoUrl: spPageContextInfo.webLogoUrl,
                permissions: spPageContextInfo.webPermMasks,
                serverRelativeUrl: spPageContextInfo.webServerRelativeUrl,
                templateName: spPageContextInfo.webTemplate,
                title: spPageContextInfo.webTitle
            }
        };
    };
    SharePointPageContextDataProvider.prototype.fetchComponentContextInfo = function (serverRelativePageUrl) {
        return this._httpClient.get(serverRelativePageUrl + '/_api/ComponentContextInfo', sp_http_1.SPHttpClient.configurations.v1)
            .then(function (response) {
            return response.json();
        }).then(function (responseObject) {
            return Promise.resolve(responseObject);
        }).catch(function (error) {
            console.log(error);
            throw error;
        });
    };
    return SharePointPageContextDataProvider;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = SharePointPageContextDataProvider;
