// @microsoft/sp-core-library/lib/Environment.js
"use strict";
var Validate_1 = require("./Validate");
var StringsProxy_1 = require("./StringsProxy");
var EnvironmentType;
(function (EnvironmentType) {
    EnvironmentType[EnvironmentType["Test"] = 0] = "Test";
    EnvironmentType[EnvironmentType["Local"] = 1] = "Local";
    EnvironmentType[EnvironmentType["SharePoint"] = 2] = "SharePoint";
    EnvironmentType[EnvironmentType["ClassicSharePoint"] = 3] = "ClassicSharePoint";
})(EnvironmentType = exports.EnvironmentType || (exports.EnvironmentType = {}));
var Environment = (function () {
    function Environment() {
    }
    Environment._initialize = function (data) {
        Validate_1.default.isNotNullOrUndefined(data, 'data');
        Environment._type = data.type;
        Validate_1.default.isNotNullOrUndefined(Environment._type, 'environmentType');
        if (!EnvironmentType.hasOwnProperty(data.type.toString())) {
            throw new Error(StringsProxy_1.default.invalidEnvironmentTypeError);
        }
    };
    Object.defineProperty(Environment, "type", {
        get: function () {
            return this._type;
        },
        enumerable: true,
        configurable: true
    });
    return Environment;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Environment;
