// @microsoft/sp-page-context/lib/test/MockPageContextData.js
"use strict";
var SPPermission_1 = require("../SPPermission");
var MockPageContextData = {
    cultureInfo: {
        currentCultureName: 'en-US',
        currentUICultureName: 'en-US',
        isRightToLeft: false
    },
    list: {
        baseTemplate: -1,
        id: '00000000-0000-4000-b000-000000000003',
        permissions: SPPermission_1.default.fullMask.value,
        serverRelativeUrl: '/sites/test/list',
        title: 'Todo List'
    },
    listItem: {
        id: -1
    },
    site: {
        absoluteUrl: 'https://wwww.contoso.com/',
        cdnPrefix: 'spfx',
        classification: 'SPX',
        correlationId: '00000000-0000-4000-b000-000000000010',
        id: '00000000-0000-4000-b000-000000000001',
        isNoScriptEnabled: false,
        recycleBinItemCount: 5,
        serverRelativeUrl: '/',
        serverRequestPath: '/test.aspx',
        sitePagesEnabled: true
    },
    user: {
        displayName: 'User 1',
        email: 'user1@contoso.com',
        isAnonymousGuestUser: false,
        isExternalGuestUser: true,
        loginName: 'i:0#.w|domain\\user'
    },
    web: {
        absoluteUrl: 'https://wwww.contoso.com/sites/test',
        id: '00000000-0000-4000-b000-000000000002',
        isAppWeb: false,
        language: -1,
        logoUrl: 'https://wwww.contoso.com/sites/test/test.jpg',
        permissions: SPPermission_1.default.fullMask.value,
        serverRelativeUrl: '/sites/test/',
        templateName: 'Blog',
        title: 'Local Workbench'
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = MockPageContextData;
