// @microsoft/sp-webpart-workbench/lib/components/workbenchCommandBar/WorkbenchCommandBar.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var React = require("react");
var office_ui_fabric_react_1 = require("office-ui-fabric-react");
var WorkbenchCommandBar_resx_1 = require("./WorkbenchCommandBar.resx");
var WorkbenchCommandBar = (function (_super) {
    __extends(WorkbenchCommandBar, _super);
    function WorkbenchCommandBar() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    WorkbenchCommandBar.prototype.render = function () {
        var items = this.props.isEditing ? [
            {
                key: 'save',
                name: WorkbenchCommandBar_resx_1.default.Save,
                icon: 'Save',
                ariaLabel: WorkbenchCommandBar_resx_1.default.SaveAltText,
                title: WorkbenchCommandBar_resx_1.default.SaveAltText,
                onClick: this.props.savePage,
                'data-automation-id': 'workbench-command-bar-save'
            },
            {
                key: 'discard',
                name: WorkbenchCommandBar_resx_1.default.Discard,
                icon: 'RevToggleKey',
                ariaLabel: WorkbenchCommandBar_resx_1.default.DiscardAltText,
                title: WorkbenchCommandBar_resx_1.default.DiscardAltText,
                onClick: this.props.clearPage,
                'data-automation-id': 'workbench-command-bar-discard'
            }
        ] : [];
        var farItems = this.props.isEditing ? [
            {
                key: 'mobilePreview',
                name: WorkbenchCommandBar_resx_1.default.Mobile,
                icon: 'CellPhone',
                ariaLabel: WorkbenchCommandBar_resx_1.default.MobleAltText,
                title: WorkbenchCommandBar_resx_1.default.MobleAltText,
                onClick: this.props.onClickMobileView,
                'data-automation-id': 'workbench-command-bar-mobile-cellphone'
            },
            {
                key: 'tabletPreview',
                name: WorkbenchCommandBar_resx_1.default.Tablet,
                icon: 'Tablet',
                ariaLabel: WorkbenchCommandBar_resx_1.default.TabletAltText,
                title: WorkbenchCommandBar_resx_1.default.TabletAltText,
                onClick: this.props.onClickTabletView,
                'data-automation-id': 'workbench-command-bar-mobile-tablet'
            },
            {
                key: 'preview',
                name: WorkbenchCommandBar_resx_1.default.Preview,
                icon: 'View',
                ariaLabel: WorkbenchCommandBar_resx_1.default.PreviewAltText,
                title: WorkbenchCommandBar_resx_1.default.PreviewAltText,
                onClick: this.props.toggleEdit,
                'data-automation-id': 'workbench-command-bar-preview'
            }
        ] : [
            {
                key: 'edit',
                name: WorkbenchCommandBar_resx_1.default.Edit,
                icon: 'Edit',
                ariaLabel: WorkbenchCommandBar_resx_1.default.EditAltText,
                title: WorkbenchCommandBar_resx_1.default.EditAltText,
                onClick: this.props.toggleEdit,
                'data-automation-id': 'workbench-command-bar-edit'
            }
        ];
        return React.createElement(office_ui_fabric_react_1.CommandBar, { isSearchBoxVisible: false, items: items, farItems: farItems });
    };
    return WorkbenchCommandBar;
}(React.Component));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = WorkbenchCommandBar;
