// @microsoft/sp-client-preview/lib/local/appBase/components/sp-canvas/toolbox/mocks/MockToolboxItemProps.js
"use strict";
var sp_core_library_1 = require("@microsoft/sp-core-library");
var CanvasControlType_1 = require("../../canvas/canvasControl/CanvasControlType");
var FEATURED_ITEM_IDS = [
    'd1d91016-032f-456d-98a4-721247c305e8_0',
    'b7dd04e1-19ce-4b24-9132-b60a1c2b910d_0',
    '275c0095-a77e-4f6d-a2a0-6a7626911518_0',
    '490d7c76-1824-45b2-9de3-676421c997fa_0',
    'daf0b71c-6de8-4ef7-b511-faae7c388708_0' 
];
var GENERIC_ITEM_ID = '00000000-0000-0000-0000-000000000000';
var MockToolboxItemProps = (function () {
    function MockToolboxItemProps() {
    }
    MockToolboxItemProps.getToolboxItemProps = function (count) {
        var toolboxItemPropMocks = [];
        var onClick = sinon.spy();
        var onChange = sinon.spy();
        for (var ct = 0; ct < count; ct++) {
            var strCt = ct.toString();
            var id = ct < 5 ? FEATURED_ITEM_IDS[ct] : GENERIC_ITEM_ID;
            var webPartManifest = {
                manifestVersion: ct,
                componentType: 'WebPart',
                id: id,
                alias: '',
                version: undefined,
                loaderConfig: undefined,
                preconfiguredEntries: undefined
            };
            var itemProps = {
                webPartData: undefined,
                webPartId: id,
                webPartManifest: webPartManifest,
                controlType: CanvasControlType_1.default.WebPartZone,
                displayMode: sp_core_library_1.DisplayMode.Edit,
                id: id,
                onChange: onChange
            };
            toolboxItemPropMocks.push({
                description: strCt + '-description',
                displayName: strCt + '-displayName',
                itemProps: itemProps,
                msIconName: strCt + '-icon',
                onClick: onClick,
                automationId: id
            });
        }
        return toolboxItemPropMocks;
    };
    return MockToolboxItemProps;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = MockToolboxItemProps;
