// @microsoft/sp-webpart-base/lib/propertyPane/propertyPaneFields/propertyPaneField/IPropertyPaneField.js
"use strict";
var PropertyPaneFieldType;
(function (PropertyPaneFieldType) {
    PropertyPaneFieldType[PropertyPaneFieldType["Custom"] = 1] = "Custom";
    PropertyPaneFieldType[PropertyPaneFieldType["CheckBox"] = 2] = "CheckBox";
    PropertyPaneFieldType[PropertyPaneFieldType["TextField"] = 3] = "TextField";
    PropertyPaneFieldType[PropertyPaneFieldType["Toggle"] = 5] = "Toggle";
    PropertyPaneFieldType[PropertyPaneFieldType["Dropdown"] = 6] = "Dropdown";
    PropertyPaneFieldType[PropertyPaneFieldType["Label"] = 7] = "Label";
    PropertyPaneFieldType[PropertyPaneFieldType["Slider"] = 8] = "Slider";
    PropertyPaneFieldType[PropertyPaneFieldType["Heading"] = 9] = "Heading";
    PropertyPaneFieldType[PropertyPaneFieldType["ChoiceGroup"] = 10] = "ChoiceGroup";
    PropertyPaneFieldType[PropertyPaneFieldType["Button"] = 11] = "Button";
    PropertyPaneFieldType[PropertyPaneFieldType["HorizontalRule"] = 12] = "HorizontalRule";
    PropertyPaneFieldType[PropertyPaneFieldType["Link"] = 13] = "Link";
})(PropertyPaneFieldType = exports.PropertyPaneFieldType || (exports.PropertyPaneFieldType = {}));
