// @microsoft/sp-build-core-tasks/lib/utilities/FileUtils.js
"use strict";
var gulpUtil = require("gulp-util");
var stream = require("stream");
var crypto = require("crypto");
function writeStringToFile(content, filename) {
    if (filename === void 0) { filename = 'tempfile'; }
    return getGulpUtilFileStream(writeStringToGulpUtilFile(content, filename));
}
exports.writeStringToFile = writeStringToFile;
function writeStringToGulpUtilFile(content, filename) {
    if (filename === void 0) { filename = 'tempfile'; }
    return new gulpUtil.File({
        contents: new Buffer(content),
        path: filename
    });
}
exports.writeStringToGulpUtilFile = writeStringToGulpUtilFile;
function getGulpUtilFileStream(file) {
    var fileStream = new stream.Readable({ objectMode: true });
    // tslint:disable-next-line:no-any
    fileStream._read = function () {
        this.push(file);
        /* tslint:disable:no-null-keyword */
        this.push(null);
        /* tslint:enable:no-null-keyword */
    };
    return fileStream;
}
exports.getGulpUtilFileStream = getGulpUtilFileStream;
function getShortHash(content) {
    return crypto.createHash('md5').update(content).digest('base64').replace('==', '');
}
exports.getShortHash = getShortHash;
