// @microsoft/sp-webpart-base/lib/test/mock/clientSideWebParts/MockWebPartContext.js
"use strict";
var MockWebPartManifestInstance_1 = require("./MockWebPartManifestInstance");
var MockWebPartContext = (function () {
    function MockWebPartContext() {
        this.domElement = document.createElement('div');
        this.manifest = new MockWebPartManifestInstance_1.default();
        this.instanceId = 'wpInstanceId';
        this.webPartTag = 'WebPart.daf0b71c-6de8-4ef7-b511-faae7c388708.0ae4973a-70e3-451e-9f1f-3f25e5455a60';
        this.spHttpClient = {};
        this.httpClient = {};
        this.pageContext = {};
        this.statusRenderer = {};
        this.eventAggregator = {};
        this.serviceScope = {};
        this.host = {};
        this.propertyPane = {
            open: function () { },
            refresh: function () { },
            isRenderedByWebPart: function () { return true; }
        };
    }
    MockWebPartContext.prototype.initialize = function (webPart) {
        webPart._internalInitialize(this);
        return webPart;
    };
    return MockWebPartContext;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = MockWebPartContext;
