// @microsoft/sp-build-core-tasks/lib/configureWebpack/ConfigureWebpackTask.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var lodash = require("lodash");
var path = require("path");
var webpack = require("webpack");
var ExternalsProcessor_1 = require("./ExternalsProcessor");
var OdspGulpTask_1 = require("./../OdspGulpTask");
var constants_1 = require("./../constants");
function getBundleId(manifest, version) {
    if (version === void 0) { version = undefined; }
    return manifest.id + "_" + (version || manifest.version);
}
exports.getBundleId = getBundleId;
var supportedWebpackLibraryTargets = ['var', 'this', 'commonjs', 'commonjs2', 'amd', 'umd'];
exports.fileLoaderExts = ['jpg', 'png', 'woff', 'eot', 'ttf', 'svg', 'gif'];
var isRunningInNpmMode = process.argv.indexOf('--npm') !== -1;
/**
 * Configures the @microsoft/gulp-core-build-webpack task with some smart defaults based on the package configuration.
 *
 * Example:
 *  IN:
 *    setConfig({
 *      webpack: <reference to webpack task>,
 *      webpackPublicPathOptions: <options for @microsoft/loader-set-webpack-public-path>,
 *      libraryName: <optional library name - set this if the library should be a global on the window>,
 *      configureExternalBundlingWebpackTask: <optional reference to ConfigureExternalBundlingWebpackTask task>
 *    })
 *    buildConfig.properties.entries = [{
 *      "entry": "./lib/App.js",
 *      "manifest": "./src/app.manifest.json",
 *      "outputPath": "dist/app.bundle.js"
 *    }]
 *    buildConfig.properties.externals = {
 *      "react-dom": "node_modules/react-dom/dist/react-dom.js",
 *      "react": "node_modules/react/dist/react.js",
 *      "flux": "node_modules/flux/dist/flux.js",
 *    }
 *
 *  OUT:
 *    set webpack config to:
 *    {
 *        context: ... root path ...,
 *        entry: { ... mapping of names to entries ... },
 *        resolveLoader: {
 *          alias: undefined,
 *          root: ... node_modules ...
 *        },
 *        resolve: {
 *          extensions: ['', '.tsx', '.ts', '.jsx', '.js', '.less', '.css', '.html'],
 *          modulesDirectories: [ ... root path + /node_modules ],
 *          alias: { }
 *        },
 *        output: {
 *          library: <library name, if provided>,
 *          libraryTarget: <library target, if specified> or 'amd',
 *          path: ... dist folder ...,
 *          publicPath: this.taskConfig.publicPath,
 *          filename: PRODUCTION ? '[name]_[chunkhash].js' : '[name].js'
 *        },
 *        externals: [ ... externals names ... ],
 *        module: {
 *          preLoaders: [],
 *          noParse: [ /\.map$/ ],
 *          loaders: [
 *            {
 *              test: /\.module.scss$/,
 *              loader:
 *                'load-themed-styles!css-loader?&modules&localIdentName=[name]__[local]___[hash:base64:5]'
 *            },
 *            {
 *              test: /\.scss$/, exclude: /\.module.scss$/,
 *              loader: 'load-themed-styles!css-loader'
 *            },
 *            {
 *              test: /\.(jpg|png|woff|eot|ttf|svg|gif)$/,
 *              loader: '@microsoft/loader-cased-file?name=[name]_[hash].[ext]'
 *            }
 *          ]
 *        },
 *        plugins: []
 *      }
 */
var ConfigureWebpackTask = (function (_super) {
    __extends(ConfigureWebpackTask, _super);
    function ConfigureWebpackTask() {
        var _this = _super.call(this) || this;
        _this.name = 'configure-webpack';
        _this.taskConfig = {
            libraryName: undefined,
            webpack: undefined,
            webpackPublicPathOptions: undefined,
            additionalConfiguration: undefined,
            configureExternalBundlingWebpackTask: undefined
        };
        _this._baseConfig = {
            module: {
                loaders: [
                    {
                        loader: '@microsoft/loader-load-themed-styles!css-loader?-minimize',
                        test: /\.css$/
                    },
                    {
                        loader: '@microsoft/loader-cased-file?name=[name:lower]_[hash].[ext]',
                        test: new RegExp(".(" + exports.fileLoaderExts.map(function (e) { return lodash.escapeRegExp(e); }).join('|') + ")((\\?|\\#).+)?$")
                    },
                    {
                        loader: 'json-loader',
                        test: /\.json$/
                    },
                    {
                        loader: 'html-loader',
                        test: /\.html$/
                    }
                ],
                noParse: [/\.map$/],
                postLoaders: [],
                preLoaders: []
            },
            resolve: {
                alias: {},
                modulesDirectories: [
                    '',
                    'node_modules',
                    'lib'
                ]
            },
            resolveLoader: {
                alias: undefined,
                root: path.join(__dirname, '..', '..', 'node_modules'),
                moduleTemplates: ['@microsoft/loader-*']
            }
        };
        // Get the webpack default configuration
        var defaultConfig = {};
        (new webpack.WebpackOptionsDefaulter()).process(defaultConfig); // tslint:disable-line:no-any
        // Include the default moduleTemplates, if there are any defaults
        if (defaultConfig.resolveLoader &&
            defaultConfig.resolveLoader.moduleTemplates &&
            !!defaultConfig.resolveLoader.moduleTemplates.length) {
            (_a = _this._baseConfig.resolveLoader.moduleTemplates).push.apply(_a, defaultConfig.resolveLoader.moduleTemplates);
        }
        return _this;
        var _a;
    }
    ConfigureWebpackTask.prototype.loadSchema = function () {
        return require('./configure-webpack.schema.json');
    };
    ConfigureWebpackTask.prototype.setConfig = function (config) {
        if (config.webpack && config.webpack.name !== 'webpack') {
            this.logWarning('Setting webpack property to non-"@microsoft/gulp-core-build-webpack"-type will ' +
                'prevent configuration');
        }
        _super.prototype.mergeConfig.call(this, config);
    };
    ConfigureWebpackTask.prototype.getKarmaConfig = function (postLoaders, externals) {
        var testAddons = {
            externals: externals,
            module: {
                loaders: [{
                        loader: 'imports?define=>false',
                        test: /sinon\.js$/
                    }],
                postLoaders: postLoaders
            },
            plugins: [new webpack.DefinePlugin({
                    DEBUG: !this.buildConfig.production,
                    UNIT_TEST: true,
                    NPM_BUILD: false,
                    BUILD_NUMBER: '\'unit_test\''
                })],
            resolve: {
                alias: {}
            }
        };
        return this._appendToBaseConfig(testAddons);
    };
    ConfigureWebpackTask.prototype.executeTask = function (gulp, completeCallback) {
        var _this = this;
        if (!this.taskConfig.webpack) {
            this.logWarning('"webpack" is not defined in the task configuration. Nothing to configure.');
            completeCallback();
            return;
        }
        if (!this.buildConfig.properties) {
            this._logNoEntriesWarning();
            this._setWebpackConfig({});
            completeCallback();
            return;
        }
        // If the library target is explicitly set, ensure that it's set to one of the valid webpack library targets.
        if (this.taskConfig.libraryTarget && supportedWebpackLibraryTargets.indexOf(this.taskConfig.libraryTarget) === -1) {
            this.logError("\"" + this.taskConfig.libraryTarget + "\" is not a valid library target type.");
            completeCallback();
            return;
        }
        // The ConfigureExternalBundlingWebpackTask will disable the second webpack task
        // when doing a dev build or in a project without localized resources.
        // This, however, breaks gulp serve. Therefore, we should ensure the webpack task is activated.
        // @todo 261553 - this should be removed when we do this refactor
        this.taskConfig.webpack.isEnabled = function () {
            return true;
        };
        /* tslint:disable:typedef variable-name */
        var StatsPlugin = require('webpack-stats-plugin');
        var VisualizerPlugin = require('webpack-visualizer-plugin');
        /* tslint:enable:typedef variable-name */
        var entries = this.properties.entries || [];
        var externalsKeys = ExternalsProcessor_1.getExternalsKeys(this);
        // Remove the specified linked externals to ensure they're bundled
        if (this.taskConfig.linkedExternalsToBundle && this.taskConfig.linkedExternalsToBundle.length) {
            for (var _i = 0, _a = this.taskConfig.linkedExternalsToBundle; _i < _a.length; _i++) {
                var linkedExternalToBundle = _a[_i];
                var linkedExternalToBundleIndex = externalsKeys.indexOf(linkedExternalToBundle);
                if (linkedExternalToBundleIndex !== -1) {
                    externalsKeys.splice(linkedExternalToBundleIndex, 1);
                }
            }
        }
        var localizedResources = this.properties.localizedResources || {};
        var localizedResourcesKeys = Object.keys(localizedResources);
        if (entries.length === 0) {
            this._logNoEntriesWarning();
        }
        this.logVerbose(entries.length + " entries specified.");
        this.logVerbose(externalsKeys.length + " externals specified.");
        this.logVerbose(localizedResourcesKeys.length + " localized resources specified.");
        localizedResourcesKeys.forEach(function (key) {
            if (externalsKeys.indexOf(key) !== -1) {
                _this.logError("Invalid Configuration: Key \"" + key + "\" is specified in both externals and localizedResources.");
            }
        });
        var webpackPublicPathOptions = lodash.clone(this.taskConfig.webpackPublicPathOptions) || {};
        if (!webpackPublicPathOptions.systemJs) {
            lodash.merge(webpackPublicPathOptions, {
                scriptPath: constants_1.default.defaultWebpackPublicPathScriptName
            });
        }
        var setWebpackPublicPath = require('@microsoft/loader-set-webpack-public-path'); // tslint:disable-line:typedef
        setWebpackPublicPath.setOptions(webpackPublicPathOptions);
        // When we loop through the config entries, record the manifests so we can use them later if we need to
        var manifests = [];
        this.properties.bundleEntries = [];
        var webpackEntries = {};
        for (var _b = 0, entries_1 = entries; _b < entries_1.length; _b++) {
            var entry = entries_1[_b];
            var entryName = path.basename(entry.outputPath, '.js');
            webpackEntries[entryName] = path.join(this.buildConfig.rootPath, entry.entry);
            var manifest = this.getManifest(entry.manifest);
            this.properties.bundleEntries.push({
                entryName: entryName,
                configEntry: entry,
                nonLocalizedEntrypointPath: path.join(this.outputDirectory, entryName + ".js"),
                manifest: manifest
            });
            if (manifest) {
                manifests.push(manifest);
            }
        }
        this.properties.webpackLibraryName = this.taskConfig.libraryName || (function () {
            if (manifests.length === 0) {
                // If there are no manifests, it doesn't appear as though we're generating a bundle that will be
                //  loaded on the page, because the module loader requires an ID, so we shouldn't need to worry about
                //  collisions in the webpackJsonp function.
                return undefined;
            }
            else if (manifests.length === 1) {
                return getBundleId(manifests[0]);
            }
            else {
                // If there are multiple manifests, this is a project that is producing multiple components. Because
                //  we need something unique for the webpack JSONP function, we'll just concatenate the IDs and then hash them.
                //  In the post-bundle steps, we'll replace the "define(<name>, ..." name with the real IDs. When we update
                //  this task to support multiple webpack configurations, we can seperate each bundle to have its own library
                //  name.
                var packageVersion_1 = _this.getPackageVersion();
                _this.properties.webpackLibraryNameIsTemporary = true;
                return _this.getChecksum(manifests.map(function (manifest) { return getBundleId(manifest, packageVersion_1); }).join(''));
            }
        })();
        var pathDirNames = this.buildConfig.rootPath.split(path.sep);
        var lastDirName = pathDirNames[pathDirNames.length - 1] === ''
            ? pathDirNames[pathDirNames.length - 2] : pathDirNames[pathDirNames.length - 1];
        var configAddons = {
            context: this.buildConfig.rootPath,
            devtool: this.buildConfig.production ? undefined : 'source-map',
            entry: webpackEntries,
            externals: externalsKeys.concat(localizedResourcesKeys),
            module: {
                preLoaders: [
                    {
                        test: /\.js$/,
                        loader: 'source-map-loader'
                    }
                ]
            },
            output: {
                chunkFilename: '[id].[name]_[chunkhash].js',
                filename: '[name].js',
                library: this.properties.webpackLibraryName,
                libraryTarget: this.taskConfig.libraryTarget || 'amd',
                path: this.outputDirectory,
                devtoolModuleFilenameTemplate: 'webpack:///../[resource-path]',
                devtoolFallbackModuleFilenameTemplate: 'webpack:///../[resource-path]?[hash]'
            },
            plugins: [
                new StatsPlugin.StatsWriterPlugin({
                    fields: null,
                    filename: path.join('..', this.buildConfig.tempFolder, "stats", lastDirName + ".stats.json")
                }),
                new VisualizerPlugin({
                    filename: path.join('..', this.buildConfig.tempFolder, "stats", lastDirName + ".stats.html")
                })
            ]
        };
        if (this.buildConfig.production) {
            configAddons.plugins.push(new webpack.optimize.UglifyJsPlugin({
                compress: {
                    dead_code: true,
                    warnings: false
                },
                mangle: true
            }));
        }
        var buildNumber = undefined;
        if (process.env.BUILD_BUILDNUMBER) {
            buildNumber = "'" + process.env.BUILD_BUILDNUMBER.toLowerCase() + "'";
        }
        configAddons.plugins.push(new webpack.DefinePlugin({
            DEBUG: !this.buildConfig.production,
            UNIT_TEST: false,
            NPM_BUILD: isRunningInNpmMode,
            BUILD_NUMBER: buildNumber
        }));
        if (externalsKeys.indexOf('office-ui-fabric-react') !== -1) {
            // IFF 'office-ui-fabric-react' is an external, include a loader to translate deep fabric references to
            // references to the top-level exports
            if (!configAddons.module) {
                configAddons.module = {};
            }
            if (!configAddons.module.preLoaders) {
                configAddons.module.preLoaders = [];
            }
            configAddons.module.preLoaders.push({
                test: /office-ui-fabric-react/,
                loader: path.join(__dirname, 'loaders', 'FabricWorkaroundLoader')
            });
        }
        var config = this._appendToBaseConfig(configAddons);
        if (this.taskConfig.configureExternalBundlingWebpackTask) {
            this.taskConfig.configureExternalBundlingWebpackTask.tinkerWithBaseWebpackConfig(config);
        }
        if (this.taskConfig.additionalConfiguration) {
            config = this.taskConfig.additionalConfiguration(config);
        }
        this._setWebpackConfig(config);
        completeCallback();
        return;
    };
    /**
     * Modify the localized webpack config to make sure the end result is correct.
     *
     * @internal
     */
    ConfigureWebpackTask.prototype.tinkerWithLocConfig = function (config) {
        config.output.libraryTarget = this.taskConfig.libraryTarget || 'amd';
    };
    ConfigureWebpackTask.prototype._appendToBaseConfig = function (config) {
        return lodash.mergeWith({}, this._baseConfig, config, function (value1, value2) {
            if (lodash.isArray(value1) && lodash.isArray(value2)) {
                return lodash.concat(value1, value2);
            }
            else {
                return undefined;
            }
        });
    };
    ConfigureWebpackTask.prototype._logNoEntriesWarning = function () {
        // @TODO: SPPPlat VSO 221536
        // this.logWarning('No entries are defined, so no bundles will be produced.');
    };
    ConfigureWebpackTask.prototype._setWebpackConfig = function (config) {
        var webpackTask = this.taskConfig.webpack;
        var sourceMapWarningSupression = /Cannot find SourceMap \'.+\'\: Error\:/;
        var suppressedWarnings = webpackTask.taskConfig.suppressWarnings || [];
        if (webpackTask) {
            webpackTask.setConfig({
                config: config,
                configPath: null,
                suppressWarnings: suppressedWarnings.concat([sourceMapWarningSupression])
            });
        }
    };
    return ConfigureWebpackTask;
}(OdspGulpTask_1.default));
exports.ConfigureWebpackTask = ConfigureWebpackTask;
