// @microsoft/sp-webpart-workbench/lib/components/canvasHost/CanvasHost.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var React = require("react");
var sp_client_preview_1 = require("@microsoft/sp-client-preview");
var sp_core_library_1 = require("@microsoft/sp-core-library");
var office_ui_fabric_react_1 = require("office-ui-fabric-react");
var PropertyPaneController_module_scss_1 = require("@microsoft/sp-webpart-base/lib/propertyPane/propertyPaneController/PropertyPaneController.module.scss");
var CanvasHost_module_scss_1 = require("./CanvasHost.module.scss");
var CanvasHost = (function (_super) {
    __extends(CanvasHost, _super);
    function CanvasHost() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CanvasHost.prototype.componentDidMount = function () {
        var _this = this;
        var canvasOptions = {
            serviceScope: this.props.store.serviceScope,
            mode: this._getDisplayMode()
        };
        this._canvas = new sp_client_preview_1.Canvas(this._canvasElement, canvasOptions);
        this.props.store.getPageReadyPromise().then(function () {
            _this._canvas.render(_this.props.store.getPageState());
            _this.props.store.setClearCallback(_this._clearPage);
            _this.props.store.setSerializeCallback(function () { return _this._canvas.serialize(); });
            _this.props.store.logSerializedStateCallback(function () {
                return _this._canvas.serialize();
            });
        });
    };
    CanvasHost.prototype.componentDidUpdate = function (props) {
        if (this._canvas) {
            this._canvas.displayMode = this._getDisplayMode();
        }
    };
    CanvasHost.prototype.render = function () {
        var _this = this;
        var rootClass = office_ui_fabric_react_1.css(CanvasHost_module_scss_1.default.canvasHost, (_a = {},
            _a[CanvasHost_module_scss_1.default.hasHeader] = this.props.hasOfficeHeader,
            _a[CanvasHost_module_scss_1.default.hasCommandBar] = this.props.hasCommandBar,
            _a));
        return (React.createElement("div", { className: rootClass },
            React.createElement("div", { className: CanvasHost_module_scss_1.default.belowHeader },
                React.createElement("div", { className: CanvasHost_module_scss_1.default.content },
                    React.createElement("div", { ref: function (canvasElement) { return _this._canvasElement = canvasElement; } })))));
        var _a;
    };
    CanvasHost.prototype._getDisplayMode = function () {
        return this.props.isEditing ? sp_core_library_1.DisplayMode.Edit : sp_core_library_1.DisplayMode.Read;
    };
    CanvasHost.prototype._clearPage = function () {
        this._canvas.render();
        var propertyPaneContainer = document.getElementById('propertyPaneContainer');
        if (propertyPaneContainer) {
            propertyPaneContainer.classList.remove(PropertyPaneController_module_scss_1.default.showPane);
            window.setTimeout(function () {
                while (propertyPaneContainer.firstChild) {
                    propertyPaneContainer.removeChild(propertyPaneContainer.firstChild);
                }
            }, 1000);
        }
    };
    return CanvasHost;
}(React.Component));
__decorate([
    office_ui_fabric_react_1.autobind
], CanvasHost.prototype, "_clearPage", null);
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = CanvasHost;
