// @microsoft/gulp-core-build/lib/tasks/GulpTask.js
"use strict";
/* tslint:disable:max-line-length */
var path = require('path');
var fs = require('fs');
var logging_1 = require('../logging');
var gutil = require('gulp-util');
var through2 = require('through2');
/* tslint:disable:typedef */
var eos = require('end-of-stream');
/* tslint:enable:typedef */
var State_1 = require('../State');
var SchemaValidator_1 = require('../jsonUtilities/SchemaValidator');
var GulpTask = (function () {
    function GulpTask() {
    }
    Object.defineProperty(GulpTask.prototype, "schema", {
        /**
         * A JSON Schema object which will be used to validate this task's configuration file
         */
        get: function () {
            return this._schema ?
                this._schema :
                this._schema = this.loadSchema();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Override this function to provide a schema which will be used to validate
     * the task's configuration file.
     */
    GulpTask.prototype.loadSchema = function () {
        return undefined;
    };
    ;
    /**
     * Shallow merges config settings into the task config.
     */
    GulpTask.prototype.setConfig = function (taskConfig) {
        /* tslint:disable:typedef */
        var objectAssign = require('object-assign');
        /* tslint:enable:typedef */
        this.taskConfig = objectAssign({}, this.taskConfig, taskConfig);
    };
    /**
     * Deep merges config settings into task config.
     */
    GulpTask.prototype.mergeConfig = function (taskConfig) {
        /* tslint:disable:typedef */
        var merge = require('lodash.merge');
        /* tslint:enable:typedef */
        this.taskConfig = merge({}, this.taskConfig, taskConfig);
    };
    /**
     * Replaces task config settings with new settings.
     */
    GulpTask.prototype.replaceConfig = function (taskConfig) {
        this.taskConfig = taskConfig;
    };
    GulpTask.prototype.onRegister = function () {
        var configFilename = this._getConfigFilePath();
        var schema = this.schema;
        var rawConfig = this._readConfigFile(configFilename, schema);
        if (rawConfig) {
            this.mergeConfig(rawConfig);
        }
    };
    GulpTask.prototype.isEnabled = function (buildConfig) {
        return (!buildConfig || !buildConfig.isRedundantBuild);
    };
    GulpTask.prototype.log = function (message) {
        logging_1.log("[" + gutil.colors.cyan(this.name) + "] " + message);
    };
    GulpTask.prototype.logVerbose = function (message) {
        logging_1.verbose("[" + gutil.colors.cyan(this.name) + "] " + message);
    };
    GulpTask.prototype.logWarning = function (message) {
        logging_1.warn("[" + gutil.colors.cyan(this.name) + "] " + message);
    };
    GulpTask.prototype.logError = function (message) {
        logging_1.error("[" + gutil.colors.cyan(this.name) + "] " + message);
    };
    GulpTask.prototype.fileError = function (filePath, line, column, errorCode, message) {
        logging_1.fileError(this.name, filePath, line, column, errorCode, message);
    };
    GulpTask.prototype.fileWarning = function (filePath, line, column, errorCode, message) {
        logging_1.fileWarning(this.name, filePath, line, column, errorCode, message);
    };
    GulpTask.prototype.getCleanMatch = function (buildConfig, taskConfig) {
        if (taskConfig === void 0) { taskConfig = this.taskConfig; }
        return this.cleanMatch;
    };
    GulpTask.prototype.execute = function (config) {
        var _this = this;
        this.buildConfig = config;
        var startTime = process.hrtime();
        logging_1.logStartSubtask(this.name);
        return new Promise(function (resolve, reject) {
            /* tslint:disable:typedef */
            var stream;
            /* tslint:enable:typedef */
            try {
                if (!_this.executeTask) {
                    throw new Error('The task subclass is missing the "executeTask" method.');
                }
                stream = _this.executeTask(_this.buildConfig.gulp, function (result) {
                    if (!result) {
                        resolve();
                    }
                    else {
                        reject(result);
                    }
                });
            }
            catch (e) {
                _this.logError(e);
                reject(e);
            }
            if (stream) {
                if (stream.then) {
                    stream.then(resolve, reject);
                }
                else if (stream.pipe) {
                    // wait for stream to end
                    eos(stream, {
                        error: true,
                        readable: stream.readable,
                        writable: stream.writable && !stream.readable
                    }, function (err) {
                        if (err) {
                            reject(err);
                        }
                        else {
                            resolve();
                        }
                    });
                    // Make sure the stream is completly read
                    stream.pipe(through2.obj(function (file, encoding, callback) {
                        'use strict';
                        callback();
                    }, function (callback) {
                        'use strict';
                        callback();
                    }));
                }
                else if (_this.executeTask.length === 1) {
                    resolve(stream);
                }
            }
            else if (_this.executeTask.length === 1) {
                resolve(stream);
            }
        })
            .then(function () {
            logging_1.logEndSubtask(_this.name, startTime);
        }, function (ex) {
            logging_1.logEndSubtask(_this.name, startTime, ex);
            throw ex;
        });
    };
    GulpTask.prototype.resolvePath = function (localPath) {
        /* tslint:disable:typedef */
        var path = require('path');
        /* tslint:enable:typedef */
        if (path.isAbsolute(localPath)) {
            return path.resolve(localPath);
        }
        return path.resolve(path.join(this.buildConfig.rootPath, localPath));
    };
    GulpTask.prototype.fileExists = function (localPath) {
        /* tslint:disable:typedef */
        var fs = require('fs');
        /* tslint:enable:typedef */
        var doesExist = false;
        var fullPath = this.resolvePath(localPath);
        try {
            doesExist = fs.statSync(fullPath).isFile();
        }
        catch (e) { }
        return doesExist;
    };
    GulpTask.prototype.copyFile = function (localSourcePath, localDestPath) {
        /* tslint:disable:typedef */
        var path = require('path');
        var fs = require('fs-extra');
        /* tslint:enable:typedef */
        var fullSourcePath = path.resolve(__dirname, localSourcePath);
        var fullDestPath = path.resolve(this.buildConfig.rootPath, (localDestPath || path.basename(localSourcePath)));
        fs.copySync(fullSourcePath, fullDestPath);
    };
    GulpTask.prototype.readJSONSync = function (localPath) {
        var fullPath = this.resolvePath(localPath);
        var result = undefined;
        /* tslint:disable:typedef */
        var fs = require('fs');
        /* tslint:enable:typedef */
        try {
            var content = fs.readFileSync(fullPath, 'utf8');
            result = JSON.parse(content);
        }
        catch (e) { }
        return result;
    };
    /**
     * Returns the path to the config file used to configure this task
     */
    GulpTask.prototype._getConfigFilePath = function () {
        return path.join(process.cwd(), 'config', this.name + ".json");
    };
    /**
     * Loads a JSON file, supports reading JSON files with comments
     */
    GulpTask.prototype._readCommentedJsonFile = function (filename) {
        return SchemaValidator_1.SchemaValidator.readCommentedJsonFile(filename);
    };
    /**
     * Helper function which loads a custom config file from disk, runs the SchemaValidator on it,
     * and then apply it to the callback defined on the IConfigurableTask
     */
    GulpTask.prototype._readConfigFile = function (filename, schema) {
        if (!fs.existsSync(filename)) {
            return undefined;
        }
        else {
            if (State_1.args['verbose']) {
                console.log("Found config file: " + path.basename(filename));
            }
            var rawData = this._readCommentedJsonFile(filename);
            if (schema) {
                SchemaValidator_1.SchemaValidator.validate(rawData, schema, filename);
            }
            return rawData;
        }
    };
    return GulpTask;
}());
exports.GulpTask = GulpTask;
