// @microsoft/sp-application-base/lib/suiteNav/O365ShellWrapper.js
"use strict";
var FlexPaneWrapper_1 = require("./FlexPaneWrapper");
var NavigationWrapper_1 = require("./NavigationWrapper");
var O365ShellWrapper = (function () {
    function O365ShellWrapper() {
        this._postRenderSignal = new Promise(function () {
            return;
        });
        this.FlexPane = new FlexPaneWrapper_1.default();
        this.Navigation = new NavigationWrapper_1.default();
    }
    O365ShellWrapper.prototype.SetData = function (shellData) {
        O365Shell.SetData(shellData);
    };
    O365ShellWrapper.prototype.Render = function (renderSettings, postRendercallback) {
        O365Shell.Render(renderSettings, postRendercallback);
    };
    O365ShellWrapper.prototype.RenderAsync = function (renderSettings, renderCallback, postRenderCallback) {
        O365Shell.RenderAsync(renderSettings, renderCallback, postRenderCallback);
    };
    O365ShellWrapper.prototype.LoadShellPlus = function () {
        O365Shell.LoadShellPlus();
    };
    O365ShellWrapper.prototype.DoesShellExist = function () {
        return typeof O365Shell !== 'undefined';
    };
    O365ShellWrapper.prototype.OnHasMember = function (memberName, callback) {
        var _this = this;
        if (this._hasMember(memberName)) {
            callback();
        }
        else {
            this._postRenderSignal.then(function () { return _this.OnHasMember(memberName, callback); });
        }
    };
    O365ShellWrapper.prototype._hasMember = function (memberName) {
        return this.DoesShellExist() && Boolean(O365Shell[memberName]);
    };
    return O365ShellWrapper;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = O365ShellWrapper;
