// @microsoft/sp-client-preview/lib/local/appBase/components/sp-canvas/canvas/canvasLayout/CanvasZonePropertyPaneControl.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var sp_webpart_base_1 = require("@microsoft/sp-webpart-base");
var CanvasControlType_1 = require("../canvasControl/CanvasControlType");
var CanvasStrings_resx_1 = require("../../loc/CanvasStrings.resx");
var CanvasZonePropertyPaneControl = (function (_super) {
    __extends(CanvasZonePropertyPaneControl, _super);
    function CanvasZonePropertyPaneControl(zone, render) {
        var _this = _super.call(this) || this;
        _this._zone = zone;
        _this._render = render;
        return _this;
    }
    CanvasZonePropertyPaneControl.prototype._internalGetPropertyPaneData = function () {
        return {
            webPartId: this._zone.index.toString(),
            title: CanvasStrings_resx_1.default.SectionPropertyPaneTitle,
            configuration: {
                pages: this._getProperyPanePages()
            },
            onPropertyPaneFieldChanged: undefined,
            onConfigurationEvent: undefined,
            onRendered: undefined,
            properties: {}
        };
    };
    CanvasZonePropertyPaneControl.prototype._internalDispose = function () {
        this._zone.dispose();
    };
    CanvasZonePropertyPaneControl.prototype._internalOnPropertyPaneFieldChanged = function (propertyPath, newValue) {
        this._zone.changeZone(newValue);
        this._render();
    };
    CanvasZonePropertyPaneControl.prototype._getCurrentColumnSize = function () {
        var sections = this._zone.sections;
        var sectionsLength = sections.length;
        var firstSectionFactor = sections[0].factor;
        var columnSize;
        if (sectionsLength === 1) {
            columnSize = firstSectionFactor === 0 ? CanvasControlType_1.default.FullWidth : CanvasControlType_1.default.OneColumn;
        }
        else if (sectionsLength === 2) {
            if (firstSectionFactor === 4) {
                columnSize = CanvasControlType_1.default.OneThirdColumnLeft;
            }
            else if (firstSectionFactor === 6) {
                columnSize = CanvasControlType_1.default.TwoColumns;
            }
            else if (firstSectionFactor === 8) {
                columnSize = CanvasControlType_1.default.OneThirdColumnRight;
            }
        }
        else if (sectionsLength === 3) {
            columnSize = CanvasControlType_1.default.ThreeColumns;
        }
        return columnSize;
    };
    CanvasZonePropertyPaneControl.prototype._getProperyPanePages = function () {
        var currentColumnSize = this._getCurrentColumnSize();
        return [{
                groups: [
                    {
                        groupName: CanvasStrings_resx_1.default.SectionPropertyPaneColumnGroupName,
                        groupFields: [
                            sp_webpart_base_1.PropertyPaneChoiceGroup('selectedLayout', {
                                options: [
                                    {
                                        key: CanvasControlType_1.default.OneColumn,
                                        checked: CanvasControlType_1.default.OneColumn === currentColumnSize,
                                        text: CanvasStrings_resx_1.default.ToolboxOneColumnPart
                                    },
                                    {
                                        key: CanvasControlType_1.default.TwoColumns,
                                        checked: CanvasControlType_1.default.TwoColumns === currentColumnSize,
                                        text: CanvasStrings_resx_1.default.ToolboxTwoColumnPart
                                    },
                                    {
                                        key: CanvasControlType_1.default.OneThirdColumnLeft,
                                        checked: CanvasControlType_1.default.OneThirdColumnLeft === currentColumnSize,
                                        text: CanvasStrings_resx_1.default.ToolboxOneThirdLeftColumnPart
                                    },
                                    {
                                        key: CanvasControlType_1.default.OneThirdColumnRight,
                                        checked: CanvasControlType_1.default.OneThirdColumnRight === currentColumnSize,
                                        text: CanvasStrings_resx_1.default.ToolboxOneThirdRightColumnPart
                                    },
                                    {
                                        key: CanvasControlType_1.default.ThreeColumns,
                                        checked: CanvasControlType_1.default.ThreeColumns === currentColumnSize,
                                        text: CanvasStrings_resx_1.default.ToolboxThreeColumnPart
                                    },
                                    {
                                        key: CanvasControlType_1.default.FullWidth,
                                        checked: CanvasControlType_1.default.FullWidth === currentColumnSize,
                                        text: CanvasStrings_resx_1.default.ToolboxFullWidthColumnPart
                                    }
                                ]
                            })
                        ]
                    }
                ]
            }];
    };
    return CanvasZonePropertyPaneControl;
}(sp_webpart_base_1.BasePropertyPaneControl));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = CanvasZonePropertyPaneControl;
