// @microsoft/gulp-core-build-typescript/lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var TypeScriptTask_1 = require("./TypeScriptTask");
var TSLintTask_1 = require("./TSLintTask");
var TextTask_1 = require("./TextTask");
var RemoveTripleSlashReferenceTask_1 = require("./RemoveTripleSlashReferenceTask");
var gulp_core_build_1 = require("@microsoft/gulp-core-build");
var ApiExtractorTask_1 = require("./ApiExtractorTask");
__export(require("./TypeScriptConfiguration"));
var TypeScriptTask_2 = require("./TypeScriptTask");
exports.TypeScriptTask = TypeScriptTask_2.TypeScriptTask;
var ApiExtractorTask_2 = require("./ApiExtractorTask");
exports.ApiExtractorTask = ApiExtractorTask_2.ApiExtractorTask;
exports.apiExtractor = new ApiExtractorTask_1.ApiExtractorTask();
exports.typescript = new TypeScriptTask_1.TypeScriptTask();
exports.tslint = new TSLintTask_1.TSLintTask();
exports.text = new TextTask_1.TextTask();
exports.removeTripleSlash = new RemoveTripleSlashReferenceTask_1.RemoveTripleSlashReferenceTask();
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:export-name
exports.default = gulp_core_build_1.parallel(exports.tslint, gulp_core_build_1.serial(exports.typescript, exports.removeTripleSlash));
