// @microsoft/sp-http/lib/spHttpClient/SPHttpClientConfiguration.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var decorators_1 = require("@microsoft/decorators");
var SPHttpClientCommonConfiguration_1 = require("./SPHttpClientCommonConfiguration");
var ODataVersion_1 = require("./ODataVersion");
var HttpClient_1 = require("../httpClient/HttpClient");
var SPHttpClientConfiguration = (function (_super) {
    __extends(SPHttpClientConfiguration, _super);
    function SPHttpClientConfiguration(flags, overrideFlags) {
        return _super.call(this, flags, overrideFlags) || this;
    }
    SPHttpClientConfiguration.prototype.overrideWith = function (sourceFlags) {
        return new SPHttpClientConfiguration(this.flags, sourceFlags);
    };
    Object.defineProperty(SPHttpClientConfiguration.prototype, "defaultSameOriginCredentials", {
        get: function () {
            return this.flags.defaultSameOriginCredentials;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SPHttpClientConfiguration.prototype, "defaultODataVersion", {
        get: function () {
            return this.flags.defaultODataVersion;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SPHttpClientConfiguration.prototype, "requestDigest", {
        get: function () {
            return this.flags.requestDigest;
        },
        enumerable: true,
        configurable: true
    });
    SPHttpClientConfiguration.prototype.initializeFlags = function () {
        _super.prototype.initializeFlags.call(this);
        this.flags.defaultSameOriginCredentials = false;
        this.flags.defaultODataVersion = ODataVersion_1.default.v3;
        this.flags.requestDigest = false;
    };
    return SPHttpClientConfiguration;
}(SPHttpClientCommonConfiguration_1.default));
__decorate([
    decorators_1.override
], SPHttpClientConfiguration.prototype, "overrideWith", null);
__decorate([
    decorators_1.override
], SPHttpClientConfiguration.prototype, "initializeFlags", null);
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = SPHttpClientConfiguration;
exports.predefinedConfigurations = {
    v1: new SPHttpClientConfiguration(HttpClient_1.default.configurations.v1, {
        jsonRequest: true,
        jsonResponse: true,
        defaultSameOriginCredentials: true,
        defaultODataVersion: ODataVersion_1.default.v4,
        requestDigest: true
    })
};
