// @microsoft/sp-client-preview/lib/local/framework/pageContext/PageContextItem.js
"use strict";
var PageContextItemState;
(function (PageContextItemState) {
    PageContextItemState[PageContextItemState["Uninitialized"] = 0] = "Uninitialized";
    PageContextItemState[PageContextItemState["Requested"] = 1] = "Requested";
    PageContextItemState[PageContextItemState["Ready"] = 2] = "Ready";
})(PageContextItemState = exports.PageContextItemState || (exports.PageContextItemState = {}));
var PageContextItem = (function () {
    function PageContextItem(pageContext) {
        this._pageContext = pageContext;
        this._state = PageContextItemState.Uninitialized;
    }
    PageContextItem.prototype.getPageContext = function () {
        return this._pageContext;
    };
    Object.defineProperty(PageContextItem.prototype, "state", {
        get: function () {
            return this._state;
        },
        set: function (value) {
            this._state = value;
        },
        enumerable: true,
        configurable: true
    });
    PageContextItem.prototype.validateLoad = function () {
        if (this._state === PageContextItemState.Ready) {
            throw new Error('The object has already been loaded');
        }
    };
    return PageContextItem;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = PageContextItem;
