// @microsoft/sp-client-preview/lib/local/common/dataAccess/DataLayerError.js
"use strict";
var DataLayerError = (function () {
    function DataLayerError(errorType, correlationId, statusCode) {
        this.errorType = errorType;
        this.correlationId = correlationId;
        this.statusCode = statusCode;
    }
    DataLayerError.prototype.toString = function () {
        return "ErrorType: " + this.errorType + ", StatusCode: " + this.statusCode + ", CorrelationId: " + this.correlationId;
    };
    return DataLayerError;
}());
exports.DataLayerError = DataLayerError;
var DataLayerErrorType;
(function (DataLayerErrorType) {
    DataLayerErrorType[DataLayerErrorType["TransportError"] = 0] = "TransportError";
    DataLayerErrorType[DataLayerErrorType["ResponseError"] = 1] = "ResponseError";
    DataLayerErrorType[DataLayerErrorType["ParseError"] = 2] = "ParseError";
    DataLayerErrorType[DataLayerErrorType["TimeOutError"] = 3] = "TimeOutError";
    DataLayerErrorType[DataLayerErrorType["UnknownError"] = 4] = "UnknownError";
})(DataLayerErrorType = exports.DataLayerErrorType || (exports.DataLayerErrorType = {}));
