// @microsoft/sp-loader/lib/test/mocks/MockSPComponentLoader.js
"use strict";
var test_1 = require("@microsoft/sp-core-library/lib/test");
var MockSPComponentLoader = (function () {
    function MockSPComponentLoader() {
    }
    Object.defineProperty(MockSPComponentLoader.prototype, "manifests", {
        get: function () {
            return test_1.TEST_UNDEFINED;
        },
        enumerable: true,
        configurable: true
    });
    MockSPComponentLoader.prototype.start = function (preloadedData) {
        return test_1.TEST_UNDEFINED;
    };
    MockSPComponentLoader.prototype.loadCss = function (url) {
        return;
    };
    MockSPComponentLoader.prototype.loadScript = function (url, options) {
        return test_1.TEST_UNDEFINED;
    };
    MockSPComponentLoader.prototype.loadComponent = function (manifest) {
        return test_1.TEST_UNDEFINED;
    };
    MockSPComponentLoader.prototype.loadComponentById = function (id, version) {
        return test_1.TEST_UNDEFINED;
    };
    MockSPComponentLoader.prototype._forceLocalManifests = function () {
        return test_1.TEST_UNDEFINED;
    };
    MockSPComponentLoader.prototype.tryGetManifestById = function (id, version) {
        return test_1.TEST_UNDEFINED;
    };
    MockSPComponentLoader.prototype.registerManifests = function (manifests) {
        return;
    };
    return MockSPComponentLoader;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = MockSPComponentLoader;
