// @microsoft/sp-loader/lib/test/TestManifests.js
"use strict";
var TestManifests = (function () {
    function TestManifests() {
    }
    return TestManifests;
}());
TestManifests.manifest = {
    manifestVersion: 2,
    componentType: 'WebPart',
    id: 'd1d91016-032f-456d-98a4-721247c305e8',
    alias: 'ImageWebpart',
    isInternal: true,
    version: '0.0.1',
    loaderConfig: {
        internalModuleBaseUrls: ['https://cdn.net/'],
        entryModuleId: 'main.bundle',
        scriptResources: {}
    }
};
TestManifests.pathDependencyName = 'myPathDependency';
TestManifests.pathDependencyConfig = {
    type: 'path',
    path: 'path/to/myDep.js'
};
TestManifests.pathDependencyFullUrl = 'https://cdn.net/path/to/myDep.js';
TestManifests.manifestWithPathDependency = {
    manifestVersion: 2,
    componentType: 'WebPart',
    id: 'd1d91016-032f-456d-98a4-721247c305e8',
    alias: 'ImageWebpart',
    isInternal: true,
    version: '0.0.1',
    loaderConfig: {
        internalModuleBaseUrls: ['https://cdn.net/'],
        entryModuleId: 'main.bundle',
        scriptResources: {
            'myPathDependency': TestManifests.pathDependencyConfig
        }
    }
};
TestManifests.localizedPathDependencyName = 'myLocalizedPathDependency';
TestManifests.localizedPathDependencyConfig = {
    type: 'localizedPath',
    defaultPath: 'localized/path/to/myDep.js'
};
TestManifests.localizedPathDependencyFullUrl = 'https://cdn.net/localized/path/to/myDep.js';
TestManifests.manifestWithLocalizedPathDependency = {
    manifestVersion: 2,
    componentType: 'WebPart',
    id: 'd1d91016-032f-456d-98a4-721247c305e8',
    alias: 'ImageWebpart',
    isInternal: true,
    version: '0.0.1',
    loaderConfig: {
        internalModuleBaseUrls: ['https://cdn.net/'],
        entryModuleId: 'main.bundle',
        scriptResources: {
            'myLocalizedPathDependency': TestManifests.localizedPathDependencyConfig
        }
    }
};
TestManifests.localizedPathWithPathsDependencyName = 'myLocalizedPathDependency';
TestManifests.localizedPathWithPathsDependencyConfig = {
    type: 'localizedPath',
    defaultPath: 'this/is/garbage',
    paths: {
        'en-us': 'localized/path/to/myDep.en.js',
        'es-es': 'localized/path/to/myDep.es.js'
    }
};
TestManifests.localizedPathWithPathsDependencyFullUrlEnUs = 'https://cdn.net/localized/path/to/myDep.en.js';
TestManifests.localizedPathWithPathsDependencyFullUrlEsEs = 'https://cdn.net/localized/path/to/myDep.es.js';
TestManifests.manifestWithLocalizedPathWithPathsDependency = {
    manifestVersion: 2,
    componentType: 'WebPart',
    id: 'd1d91016-032f-456d-98a4-721247c305e8',
    alias: 'ImageWebpart',
    isInternal: true,
    version: '0.0.1',
    loaderConfig: {
        internalModuleBaseUrls: ['https://cdn.net/'],
        entryModuleId: 'main.bundle',
        scriptResources: {
            'myLocalizedPathDependency': TestManifests.localizedPathWithPathsDependencyConfig
        }
    }
};
TestManifests.componentDependencyName = 'myComponentDependency';
TestManifests.componentDependencyId = '7263c7d0-1d6a-45ec-8d85-d4d1d234171b';
TestManifests.componentDependencyVersion = '1.0.0';
TestManifests.componentDependencyConfig = {
    type: 'component',
    id: TestManifests.componentDependencyId,
    version: TestManifests.componentDependencyVersion
};
TestManifests.failoverPath = 'https://my.domain.invalid/myComponentDependency-1.0.0.js';
TestManifests.componentDependencyConfigWithFailoverPath = {
    type: 'component',
    id: TestManifests.componentDependencyId,
    version: TestManifests.componentDependencyVersion,
    failoverPath: TestManifests.failoverPath
};
TestManifests.manifestWithComponentDependency = {
    manifestVersion: 2,
    componentType: 'WebPart',
    id: 'd1d91016-032f-456d-98a4-721247c305e8',
    alias: 'ImageWebpart',
    isInternal: true,
    version: '0.0.1',
    loaderConfig: {
        internalModuleBaseUrls: ['https://cdn.net/'],
        entryModuleId: 'main.bundle',
        scriptResources: {
            'myComponentDependency': TestManifests.componentDependencyConfig
        }
    }
};
TestManifests.componentDependencyManifest = {
    manifestVersion: 2,
    componentType: 'WebPart',
    id: TestManifests.componentDependencyId,
    alias: 'componentDependency',
    isInternal: true,
    version: TestManifests.componentDependencyVersion,
    loaderConfig: {
        internalModuleBaseUrls: ['https://cdn.net/'],
        entryModuleId: 'main.bundle',
        scriptResources: {}
    }
};
TestManifests.manifestWithComponentDependencyWithFailoverPath = {
    manifestVersion: 2,
    componentType: 'WebPart',
    id: 'd1d91016-032f-456d-98a4-721247c305e8',
    alias: 'ImageWebpart',
    isInternal: true,
    version: '0.0.1',
    loaderConfig: {
        internalModuleBaseUrls: ['https://cdn.net/'],
        entryModuleId: 'main.bundle',
        scriptResources: {
            'myComponentDependency': TestManifests.componentDependencyConfigWithFailoverPath
        }
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = TestManifests;
