// @microsoft/sp-webpart-base/lib/core/test/TestPersonCardWebPart.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var BaseClientSideWebPart_1 = require("../BaseClientSideWebPart");
var test_1 = require("@microsoft/sp-core-library/lib/test");
var TestPersonCardWebPart = (function (_super) {
    __extends(TestPersonCardWebPart, _super);
    function TestPersonCardWebPart() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TestPersonCardWebPart.prototype.render = function () {
        return;
    };
    Object.defineProperty(TestPersonCardWebPart.prototype, "propertiesMetadata", {
        get: function () {
            return {
                'photo': { isImageSource: true },
                'websites[*].description': { isHtmlString: true },
                'websites[*].url': { isLink: true },
                'manager.photo': { isImageSource: true },
                'manager.websites[*].description': { isHtmlString: true },
                'manager.websites[*].url': { isLink: true }
            };
        },
        enumerable: true,
        configurable: true
    });
    return TestPersonCardWebPart;
}(BaseClientSideWebPart_1.default));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = TestPersonCardWebPart;
var runtimePropertyBag = {
    name: 'John Doe',
    photo: '//path/johndoe.png',
    websites: [
        {
            title: 'John Doe\'s Homepage',
            description: 'A <b>bold</b> description',
            url: 'http://johndoe.com'
        },
        {
            title: 'John Doe\'s blog',
            url: 'blog.johndoe.com'
        }
    ],
    manager: {
        name: 'Mr Manager',
        photo: '//path/default.png',
        websites: [
            {
                title: 'Manager Homepage',
                description: '<div>Safe part</div><script>non-safe part</script>',
                url: '//manager/page'
            }
        ]
    }
};
var serializedPropertyBag = {
    name: runtimePropertyBag.name,
    photo: test_1.TEST_UNDEFINED,
    websites: [
        {
            title: runtimePropertyBag.websites[0].title,
            description: test_1.TEST_UNDEFINED,
            url: test_1.TEST_UNDEFINED
        },
        {
            title: runtimePropertyBag.websites[1].title,
            url: test_1.TEST_UNDEFINED
        }
    ],
    manager: {
        name: runtimePropertyBag.manager.name,
        websites: [
            {
                title: runtimePropertyBag.manager.websites[0].title,
                description: test_1.TEST_UNDEFINED,
                url: test_1.TEST_UNDEFINED
            }
        ]
    }
};
var serializedServerContent = {
    htmlStrings: {
        'websites[0].description': runtimePropertyBag.websites[0].description,
        'manager.websites[0].description': runtimePropertyBag.manager.websites[0].description
    },
    imageSources: {
        'photo': runtimePropertyBag.photo,
        'manager.photo': runtimePropertyBag.manager.photo
    },
    links: {
        'websites[0].url': runtimePropertyBag.websites[0].url,
        'websites[1].url': runtimePropertyBag.websites[1].url,
        'manager.websites[0].url': runtimePropertyBag.manager.websites[0].url
    }
};
var deserializedServerContent = {
    htmlStrings: {
        'websites[0].description': 'A <b>bold</b> description',
        'manager.websites[0].description': '<div>Safe part</div>'
    },
    imageSources: {
        'photo': '//cdnpath/johndoe.png',
        'manager.photo': '//cdnpath/default.png'
    },
    links: {
        'websites[0].url': 'http://johndoe.com',
        'websites[1].url': '//johndoe.com/blog',
        'manager.websites[0].url': '//manager/page'
    }
};
var deserializedPropertyBag = {
    name: serializedPropertyBag.name,
    photo: test_1.TEST_UNDEFINED,
    websites: [
        {
            title: serializedPropertyBag.websites[0].title,
            description: 'Tampered description',
            url: test_1.TEST_UNDEFINED
        },
        {
            title: serializedPropertyBag.websites[1].title,
            url: test_1.TEST_UNDEFINED
        }
    ],
    manager: {
        name: serializedPropertyBag.name,
        photo: test_1.TEST_UNDEFINED,
        websites: [
            {
                title: serializedPropertyBag.websites[0].title,
                description: test_1.TEST_UNDEFINED,
                url: test_1.TEST_UNDEFINED
            }
        ]
    }
};
var fixedPropertyBag = {
    name: deserializedPropertyBag.name,
    photo: deserializedServerContent.imageSources['photo'],
    websites: [
        {
            title: deserializedPropertyBag.websites[0].title,
            description: deserializedServerContent.htmlStrings['websites[0].description'],
            url: deserializedServerContent.links['websites[0].url']
        },
        {
            title: deserializedPropertyBag.websites[1].title,
            url: deserializedServerContent.links['websites[1].url']
        }
    ],
    manager: {
        name: deserializedPropertyBag.manager.name,
        photo: deserializedServerContent.imageSources['manager.photo'],
        websites: [
            {
                title: deserializedPropertyBag.manager.websites[0].title,
                description: deserializedServerContent.htmlStrings['manager.websites[0].description'],
                url: deserializedServerContent.links['manager.websites[0].url']
            }
        ]
    }
};
exports.testPersonCardWebPartMocks = {
    runtimePropertyBag: runtimePropertyBag,
    serializedPropertyBag: serializedPropertyBag,
    serializedServerContent: serializedServerContent,
    deserializedServerContent: deserializedServerContent,
    deserializedPropertyBag: deserializedPropertyBag,
    fixedPropertyBag: fixedPropertyBag
};
