// @microsoft/sp-webpart-base/lib/core/test/WebPartMocks.js
"use strict";
var sp_core_library_1 = require("@microsoft/sp-core-library");
var test_1 = require("@microsoft/sp-core-library/lib/test");
var sp_page_context_1 = require("@microsoft/sp-page-context");
var EventAggregator_1 = require("../events/EventAggregator");
var MockWebPartManifest = (function () {
    function MockWebPartManifest() {
        this.manifestVersion = 2;
        this.id = 'wpid';
        this.componentType = 'WebPart';
        this.alias = 'MockWebPart';
        this.version = '1.0.0';
        this.preconfiguredEntries = [{
                title: { default: 'wpName' },
                description: { default: 'wpDescription' },
                officeFabricIconFontName: 'wpIcon',
                iconImageUrl: 'wpIcon',
                groupId: 'group',
                properties: {}
            }];
        this.loaderConfig = test_1.TEST_UNDEFINED;
    }
    return MockWebPartManifest;
}());
exports.MockWebPartManifest = MockWebPartManifest;
var MockWebPartManagerContext = (function () {
    function MockWebPartManagerContext() {
        this.domElement = document.createElement('div');
        this.instanceId = 'wpInstanceId';
        this.manifest = new MockWebPartManifest();
        this.webPartData = test_1.TEST_UNDEFINED;
        this.displayMode = sp_core_library_1.DisplayMode.Read;
        this.host = test_1.TEST_UNDEFINED;
    }
    return MockWebPartManagerContext;
}());
exports.MockWebPartManagerContext = MockWebPartManagerContext;
var MockPreloadedPageData = (function () {
    function MockPreloadedPageData() {
        this.appId = '00000000-0000-0000-0000-000000000000';
        this.webServerRelativeUrl = 'https://localhost:4321';
        this.webAbsoluteUrl = 'https://localhost:4321';
    }
    return MockPreloadedPageData;
}());
exports.MockPreloadedPageData = MockPreloadedPageData;
var MockWebPartHost = (function () {
    function MockWebPartHost() {
        this.pageContext = new sp_page_context_1.PageContext(test_1.TEST_UNDEFINED);
        this.httpClient = test_1.TEST_UNDEFINED;
        this.spHttpClient = test_1.TEST_UNDEFINED;
        this.eventAggregator = EventAggregator_1.default;
        this.statusRenderer = new MockClientSideWebPartStatusRenderer();
        this.serviceScope = test_1.TEST_UNDEFINED;
        this.renderPropertyPane = function (props) { };
    }
    return MockWebPartHost;
}());
exports.MockWebPartHost = MockWebPartHost;
var MockClientSideWebPartStatusRenderer = (function () {
    function MockClientSideWebPartStatusRenderer() {
        this.displayCalled = false;
        this.clearCalled = false;
        this.errorCalled = false;
        this.clearErrorCalled = false;
    }
    MockClientSideWebPartStatusRenderer.prototype.displayLoadingIndicator = function (domElement) {
        this.displayCalled = true;
    };
    MockClientSideWebPartStatusRenderer.prototype.clearLoadingIndicator = function (domElement) {
        this.clearCalled = true;
    };
    MockClientSideWebPartStatusRenderer.prototype.renderError = function (domElement, error) {
        this.errorCalled = true;
    };
    MockClientSideWebPartStatusRenderer.prototype.clearError = function (domElement) {
        this.clearErrorCalled = true;
    };
    return MockClientSideWebPartStatusRenderer;
}());
exports.MockClientSideWebPartStatusRenderer = MockClientSideWebPartStatusRenderer;
