// @microsoft/sp-webpart-workbench/lib/stores/PageStore.js
"use strict";
var PageStore = (function () {
    function PageStore(serviceScope) {
        var _this = this;
        this._PAGE_DATA_KEY = 'sp_webpart_workbench_state';
        this._pageReadyPromise = new Promise(function (resolve) {
            _this._resolvePageReady = resolve;
        });
        this._serviceScope = serviceScope;
        this.clearPageState = this.clearPageState.bind(this);
        this.setSerializeCallback = this.setSerializeCallback.bind(this);
        this.setClearCallback = this.setClearCallback.bind(this);
        this.savePageState = this.savePageState.bind(this);
        this.getPageState = this.getPageState.bind(this);
        this.getPageReadyPromise = this.getPageReadyPromise.bind(this);
        this.setPageReady = this.setPageReady.bind(this);
        window.setInterval(this.savePageState, 3 * 1000);
    }
    Object.defineProperty(PageStore.prototype, "serviceScope", {
        get: function () {
            return this._serviceScope;
        },
        enumerable: true,
        configurable: true
    });
    PageStore.prototype.setSerializeCallback = function (callback) {
        this._serializePage = callback;
    };
    PageStore.prototype.logSerializedStateCallback = function (callback) {
        if (window['SP']) {
            window['SP'].logSerializedPageState = callback;
        }
    };
    PageStore.prototype.setClearCallback = function (callback) {
        this._clearPage = callback;
    };
    PageStore.prototype.savePageState = function () {
        if (this._serializePage) {
            sessionStorage.setItem(this._PAGE_DATA_KEY, this._serializePage());
        }
    };
    PageStore.prototype.clearPageState = function () {
        sessionStorage.removeItem(this._PAGE_DATA_KEY);
        if (this._clearPage) {
            this._clearPage();
        }
    };
    PageStore.prototype.getPageState = function () {
        return sessionStorage.getItem(this._PAGE_DATA_KEY);
    };
    PageStore.prototype.getPageReadyPromise = function () {
        return this._pageReadyPromise;
    };
    PageStore.prototype.setPageReady = function () {
        this._resolvePageReady();
    };
    return PageStore;
}());
exports.PageStore = PageStore;
