// @microsoft/sp-client-preview/lib/local/appBase/components/sp-rte/formattingbar/SPRteFormattingBar.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var office_ui_fabric_react_1 = require("office-ui-fabric-react");
var React = require("react");
require("./SPRteFormattingBar.scss");
var office_ui_fabric_react_2 = require("office-ui-fabric-react");
var SPRteFormattingBar = (function (_super) {
    __extends(SPRteFormattingBar, _super);
    function SPRteFormattingBar(props) {
        var _this = _super.call(this, props) || this;
        _this.state = {
            isFaded: false
        };
        return _this;
    }
    SPRteFormattingBar.prototype.componentDidMount = function () {
        this._domNode.addEventListener('keydown', this.props.onKeyDown);
    };
    SPRteFormattingBar.prototype.componentWillReceiveProps = function (nextProps) {
        if (this.props.position.top !== nextProps.position.top) {
            this.setState({
                isFaded: true
            });
        }
    };
    SPRteFormattingBar.prototype.componentDidUpdate = function () {
        var _this = this;
        if (this.state.isFaded) {
            window.setTimeout(function () {
                _this.setState({
                    isFaded: false
                });
            }, 300);
        }
    };
    SPRteFormattingBar.prototype.render = function () {
        var style = this.props.position;
        var className = office_ui_fabric_react_1.css('SPRteFormattingBar', 'rteFlyout-visible', 'stack-panel-class', 'ms-u-slideUpIn10', {
            'faded': this.state.isFaded
        });
        return (React.createElement("div", { role: 'toolbar', "aria-label": 'Formatting', className: className, ref: this._resolveRef('_domNode'), style: style, "data-automation-id": 'toolbar' },
            React.createElement(office_ui_fabric_react_1.FocusZone, { ref: this._resolveRef('_focusZone'), isCircularNavigation: true }, this.props.buttons)));
    };
    SPRteFormattingBar.prototype.focus = function () {
        this._focusZone.focus();
    };
    return SPRteFormattingBar;
}(office_ui_fabric_react_2.BaseComponent));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = SPRteFormattingBar;
