// office-ui-fabric-react/lib-amd/components/Spinner/Spinner.Props.js
define(["require", "exports"], function (require, exports) {
    "use strict";
    var SpinnerSize;
    (function (SpinnerSize) {
        /**
         * 12px Spinner diameter
         */
        SpinnerSize[SpinnerSize["xSmall"] = 0] = "xSmall";
        /**
         * 16px Spinner diameter
         */
        SpinnerSize[SpinnerSize["small"] = 1] = "small";
        /**
         * 20px Spinner diameter
         */
        SpinnerSize[SpinnerSize["medium"] = 2] = "medium";
        /**
         * 28px Spinner diameter
         */
        SpinnerSize[SpinnerSize["large"] = 3] = "large";
    })(SpinnerSize = exports.SpinnerSize || (exports.SpinnerSize = {}));
    /**
     * @deprecated
     * Deprecated and will be removed at >= 2.0.0. Use SpinnerSize instead.
     */
    var SpinnerType;
    (function (SpinnerType) {
        /**
         * Deprecated and will be removed at >= 2.0.0. Use SpinnerSize.medium instead.
         */
        SpinnerType[SpinnerType["normal"] = 0] = "normal";
        /**
         * Deprecated and will be removed at >= 2.0.0. Use SpinnerSize.large instead.
         */
        SpinnerType[SpinnerType["large"] = 1] = "large";
    })(SpinnerType = exports.SpinnerType || (exports.SpinnerType = {}));
});
