// mobservable/lib/index.js
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var core = require('./core');
var dnode_1 = require('./dnode');
var utils_1 = require('./utils');
var extras_1 = require('./extras');
var simpleeventemitter_1 = require('./simpleeventemitter');
__export(require('./interfaces'));
var core_1 = require('./core');
exports.isObservable = core_1.isObservable;
exports.isObservableObject = core_1.isObservableObject;
exports.isObservableArray = core_1.isObservableArray;
exports.isObservableMap = core_1.isObservableMap;
exports.observable = core_1.observable;
exports.extendObservable = core_1.extendObservable;
exports.asReference = core_1.asReference;
exports.asFlat = core_1.asFlat;
exports.asStructure = core_1.asStructure;
exports.observe = core_1.observe;
exports.autorun = core_1.autorun;
exports.autorunUntil = core_1.autorunUntil;
exports.autorunAsync = core_1.autorunAsync;
exports.expr = core_1.expr;
exports.toJSON = core_1.toJSON;
exports.isReactive = core_1.isObservable;
exports.map = core_1.map;
exports.fastArray = core_1.fastArray;
exports.makeReactive = core_1.observable;
exports.extendReactive = core_1.extendObservable;
exports.observeUntil = core_1.autorunUntil;
exports.observeAsync = core_1.autorunAsync;
var transform_1 = require('./transform');
exports.createTransformer = transform_1.createTransformer;
var dnode_2 = require('./dnode');
exports.untracked = dnode_2.untracked;
exports.transaction = dnode_2.transaction;
var observablemap_1 = require('./observablemap');
exports.ObservableMap = observablemap_1.ObservableMap;
exports._ = {
    isComputingView: dnode_1.isComputingView,
    quickDiff: utils_1.quickDiff
};
exports.extras = {
    getDNode: extras_1.getDNode,
    getDependencyTree: extras_1.getDependencyTree,
    getObserverTree: extras_1.getObserverTree,
    trackTransitions: extras_1.trackTransitions,
    SimpleEventEmitter: simpleeventemitter_1.default,
    withStrict: core.withStrict
};
