// @microsoft/sp-client-preview/lib/local/framework/shell/ClientSideApplication.js
"use strict";
var SuiteNavManagerConfiguration_1 = require("./SuiteNavManagerConfiguration");
var ClientSideApplication = (function () {
    function ClientSideApplication() {
        this._shell = undefined;
        this._domElement = undefined;
        this['__type'] = 'ClientSideApplication';
    }
    Object.defineProperty(ClientSideApplication.prototype, "shell", {
        get: function () {
            return this._shell;
        },
        set: function (value) {
            throw new Error('The property cannot be assigned because it is read-only');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ClientSideApplication.prototype, "domElement", {
        get: function () {
            return this._domElement;
        },
        set: function (value) {
            throw new Error('The property cannot be assigned because it is read-only');
        },
        enumerable: true,
        configurable: true
    });
    ClientSideApplication.prototype.load = function (shell) {
        this._shell = shell;
        this.onLoad();
    };
    ClientSideApplication.prototype.onLoad = function () {
    };
    ClientSideApplication.prototype.render = function (domElement) {
        this._domElement = domElement;
        this.onRender();
    };
    ClientSideApplication.prototype.onRender = function () {
    };
    ClientSideApplication.prototype.suiteNavConfiguration = function () {
        return new SuiteNavManagerConfiguration_1.default(this._shell.appPageContext);
    };
    return ClientSideApplication;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = ClientSideApplication;
