// @microsoft/sp-client-preview/lib/local/framework/shell/SuiteNavManagerConfiguration.js
"use strict";
var SuiteNavLinkIds_1 = require("../../common/suiteNav/models/SuiteNavLinkIds");
var SuiteNavManagerConfiguration = (function () {
    function SuiteNavManagerConfiguration(pageContext) {
        this._pageContext = pageContext;
        this._currentUICultureName = this._pageContext.core.currentUICultureName;
        this._webServerRelativeUrl = this._pageContext.core.webServerRelativeUrl;
        this._o365ShellRenderSettings = {
            disableDelayLoad: false,
            disableShellPlus: false,
            layout: 'Mouse',
            top: undefined
        };
        this._hamburgerMenuCallback = undefined;
        this._suiteNavPostRenderCallback = undefined;
        this._systemUserKey = this._pageContext.core.systemUserKey;
        this._siteClientTag = this._pageContext.core.siteClientTag;
        this._isSuiteNavDisabled = this._shouldDisableSuiteNav();
        if (this._pageContext.core.menuData) {
            this._settingsLinks = this._pageContext.core.menuData['SettingsData'];
            this._signoutUrl = this._pageContext.core.menuData['SignOutUrl'];
        }
    }
    SuiteNavManagerConfiguration.prototype.isSuiteNavDisabled = function () {
        return this._isSuiteNavDisabled;
    };
    SuiteNavManagerConfiguration.prototype.disableSuiteNav = function () {
        this._isSuiteNavDisabled = true;
    };
    SuiteNavManagerConfiguration.prototype.modifySuiteNavData = function (suiteNavData) {
        suiteNavData.CurrentMainLinkElementID = SuiteNavLinkIds_1.default.SITES;
        suiteNavData.CurrentWorkloadSettingsSubLinks =
            this._settingsLinks && this._settingsLinks.length > 0 ? this._settingsLinks : undefined;
        suiteNavData.FeedbackLink = undefined;
        if (suiteNavData.SignOutLink) {
            suiteNavData.SignOutLink.Url = this._signoutUrl;
        }
        if (suiteNavData.HelpLink) {
            suiteNavData.HelpLink.Url = 'https://go.microsoft.com/fwlink/p/?linkid=827181';
        }
        return suiteNavData;
    };
    ;
    Object.defineProperty(SuiteNavManagerConfiguration.prototype, "systemUserKey", {
        get: function () {
            return this._systemUserKey || '';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SuiteNavManagerConfiguration.prototype, "cacheToken", {
        get: function () {
            var clientTag = this._siteClientTag;
            if (clientTag !== undefined) {
                return clientTag.substring(0, clientTag.indexOf('$$'));
            }
            return '';
        },
        enumerable: true,
        configurable: true
    });
    ;
    Object.defineProperty(SuiteNavManagerConfiguration.prototype, "o365ShellRenderSettings", {
        get: function () {
            return this._o365ShellRenderSettings;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SuiteNavManagerConfiguration.prototype, "currentUICultureName", {
        get: function () {
            return this._currentUICultureName;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SuiteNavManagerConfiguration.prototype, "webServerRelativeUrl", {
        get: function () {
            return this._webServerRelativeUrl;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SuiteNavManagerConfiguration.prototype, "hamburgerCallback", {
        get: function () {
            return this._hamburgerMenuCallback;
        },
        set: function (hamburgerCallback) {
            this._hamburgerMenuCallback = hamburgerCallback;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SuiteNavManagerConfiguration.prototype, "suiteNavPostRenderCallback", {
        get: function () {
            return this._suiteNavPostRenderCallback;
        },
        set: function (suiteNavPostRenderCallback) {
            this._suiteNavPostRenderCallback = suiteNavPostRenderCallback;
        },
        enumerable: true,
        configurable: true
    });
    SuiteNavManagerConfiguration.prototype._shouldDisableSuiteNav = function () {
        return (window.location.hostname === 'localhost' ||
            this._pageContext.urlQueryParameters.getValue('disableSuiteNav') === 'true') ?
            true :
            false;
    };
    return SuiteNavManagerConfiguration;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = SuiteNavManagerConfiguration;
