// @microsoft/sp-webpart-workbench/lib/stores/PageStore.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var office_ui_fabric_react_1 = require("office-ui-fabric-react");
var PageStore = (function () {
    function PageStore(serviceScope) {
        var _this = this;
        this._PAGE_DATA_KEY = 'sp_webpart_workbench_state';
        this._pageReadyPromise = new Promise(function (resolve) {
            _this._resolvePageReady = resolve;
        });
        this._serviceScope = serviceScope;
        window.setInterval(this.savePageState, 3 * 1000);
    }
    Object.defineProperty(PageStore.prototype, "serviceScope", {
        get: function () {
            return this._serviceScope;
        },
        enumerable: true,
        configurable: true
    });
    PageStore.prototype.setSerializeCallback = function (callback) {
        this._serializePage = callback;
    };
    PageStore.prototype.logSerializedStateCallback = function (callback) {
        if (window['SP']) {
            window['SP'].logSerializedPageState = callback;
        }
    };
    PageStore.prototype.setClearCallback = function (callback) {
        this._clearPage = callback;
    };
    PageStore.prototype.savePageState = function () {
        if (this._serializePage) {
            sessionStorage.setItem(this._PAGE_DATA_KEY, this._serializePage());
        }
    };
    PageStore.prototype.clearPageState = function () {
        sessionStorage.removeItem(this._PAGE_DATA_KEY);
        if (this._clearPage) {
            this._clearPage();
        }
    };
    PageStore.prototype.getPageState = function () {
        return sessionStorage.getItem(this._PAGE_DATA_KEY);
    };
    PageStore.prototype.getPageReadyPromise = function () {
        return this._pageReadyPromise;
    };
    PageStore.prototype.setPageReady = function () {
        this._resolvePageReady();
    };
    return PageStore;
}());
__decorate([
    office_ui_fabric_react_1.autobind
], PageStore.prototype, "setSerializeCallback", null);
__decorate([
    office_ui_fabric_react_1.autobind
], PageStore.prototype, "setClearCallback", null);
__decorate([
    office_ui_fabric_react_1.autobind
], PageStore.prototype, "savePageState", null);
__decorate([
    office_ui_fabric_react_1.autobind
], PageStore.prototype, "clearPageState", null);
__decorate([
    office_ui_fabric_react_1.autobind
], PageStore.prototype, "getPageState", null);
__decorate([
    office_ui_fabric_react_1.autobind
], PageStore.prototype, "getPageReadyPromise", null);
__decorate([
    office_ui_fabric_react_1.autobind
], PageStore.prototype, "setPageReady", null);
exports.PageStore = PageStore;
