// @microsoft/sp-loader/lib/utilities/telemetryConstants.js
"use strict";
exports.startQosScenarioName = 'SPComponentLoader.start';
exports.platformFailedToLoadFailure = 'PlatformFailedToLoad';
exports.applicationFailedToLoadFailure = 'ApplicationFailedToLoad';
exports.shellStartFailure = 'SPShell.Start';
exports.missingPlatformManifestFailure = 'MissingPlatformManifest';
exports.missingApplicationManifestFailure = 'MissingApplicationManifest';
exports.missingApplicationIdFailure = 'MissingApplicationId';
exports.invalidPreloadedDataFailure = 'InvalidPreloadedData';
exports.loadComponentQosScenarioName = 'SPComponentLoader.loadComponent';
exports.loadComponentImplQosScenarioName = 'SPComponentLoader._loadComponentImpl';
exports.loadPathDependencyQosScenarioName = 'SPComponentLoader.loadPathDependency';
exports.emptyComponentErrorTagName = 'EmptyComponent';
exports.configureSystemJsErrorTagName = 'ConfigureSystemJs';
exports.loadComponentDependenciesErrorTagName = 'LoadComponentDependencies';
exports.loadPathDependenciesErrorTagName = 'LoadPathDependencies';
exports.spStarterLogSource = { id: 'SPStarter.start' };
exports.startLogSource = { id: 'SPComponentLoader.start' };
exports.loadComponentLogSource = { id: 'SPComponentLoader.loadComponent' };
exports.loadScriptLogSource = { id: 'SPComponentLoader.loadScript' };
exports.resolveAddressLogSource = { id: 'resolveAddress' };
exports.componentStoreLogSource = { id: 'ComponentStore' };
exports.manifestStoreLogSource = { id: 'ManifestStore' };
