// @microsoft/sp-loader/lib/DeveloperTools/Stores/DeveloperToolsConsoleStore.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var BaseStore_1 = require("./BaseStore");
var ManifestManager_1 = require("./../Components/DeveloperModules/ManifestManager/ManifestManager");
var PerformanceDisplay_1 = require("./../Components/DeveloperModules/PerformanceDisplay/PerformanceDisplay");
var TraceDisplay_1 = require("./../Components/DeveloperModules/TraceDisplay/TraceDisplay");
var DeveloperToolsTab_1 = require("./../Models/DeveloperToolsTab");
var ManifestManager_resx_1 = require("./../Components/DeveloperModules/ManifestManager/ManifestManager.resx");
var PerformanceDisplay_resx_1 = require("./../Components/DeveloperModules/PerformanceDisplay/PerformanceDisplay.resx");
var TraceDisplay_resx_1 = require("./../Components/DeveloperModules/TraceDisplay/TraceDisplay.resx");
var DEFAULT_HEIGHT = 1 / 4;
var MIN_HEIGHT = 1 / 6;
var MAX_HEIGHT = 3 / 4;
var DeveloperToolsConsoleStore = (function (_super) {
    __extends(DeveloperToolsConsoleStore, _super);
    function DeveloperToolsConsoleStore() {
        var _this = _super.call(this) || this;
        _this._height = DEFAULT_HEIGHT;
        _this._visible = false;
        _this._selectedTabId = 0;
        _this._tabs = [
            new DeveloperToolsTab_1.default({
                title: ManifestManager_resx_1.default.tabTitle,
                component: ManifestManager_1.default
            }),
            new DeveloperToolsTab_1.default({
                title: PerformanceDisplay_resx_1.default.TabTitle,
                component: PerformanceDisplay_1.default
            }),
            new DeveloperToolsTab_1.default({
                title: TraceDisplay_resx_1.default.TabTitle,
                component: TraceDisplay_1.default
            })
        ];
        return _this;
    }
    Object.defineProperty(DeveloperToolsConsoleStore, "instance", {
        get: function () {
            if (!this._instance) {
                this._instance = new DeveloperToolsConsoleStore();
            }
            return this._instance;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DeveloperToolsConsoleStore.prototype, "height", {
        get: function () {
            return this._height;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DeveloperToolsConsoleStore.prototype, "visible", {
        get: function () {
            return this._visible;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DeveloperToolsConsoleStore.prototype, "selectedTabId", {
        get: function () {
            return this._selectedTabId;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DeveloperToolsConsoleStore.prototype, "tabs", {
        get: function () {
            return this._tabs;
        },
        enumerable: true,
        configurable: true
    });
    DeveloperToolsConsoleStore.prototype.resizeConsole = function (height) {
        var newHeight = this._height;
        if (height >= MAX_HEIGHT) {
            newHeight = MAX_HEIGHT;
        }
        else if (height <= MIN_HEIGHT) {
            newHeight = MIN_HEIGHT;
        }
        else {
            newHeight = height;
        }
        if (newHeight !== this._height) {
            this._height = newHeight;
            this.emitChange();
        }
    };
    DeveloperToolsConsoleStore.prototype.showHideConsole = function (show) {
        if (this._visible !== show) {
            this._visible = show;
            this.emitChange();
        }
    };
    DeveloperToolsConsoleStore.prototype.registerTab = function (tab) {
        this._tabs.push(tab);
        if (this._tabs.length === 1) {
            this._selectedTabId = 0;
        }
        this.emitChange();
    };
    return DeveloperToolsConsoleStore;
}(BaseStore_1.default));
exports.DeveloperToolsConsoleStore = DeveloperToolsConsoleStore;
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = DeveloperToolsConsoleStore.instance;
