// @microsoft/sp-loader/lib/SPStarter.js
"use strict";
var developerToolsLoader = require('./DeveloperTools/DeveloperToolsLoader');
var SPSystemJsComponentLoader_1 = require('./SPSystemJsComponentLoader');
var sp_core_library_1 = require('@microsoft/sp-core-library');
var sp_telemetry_1 = require('@ms/sp-telemetry');
var SPStarter = (function () {
    function SPStarter() {
    }
    SPStarter.start = function (preloadedData) {
        if (preloadedData && preloadedData.spPageContextInfo) {
            SPStarter._initializeEnvironment();
            sp_core_library_1.Session._initialize({ applicationId: preloadedData.spPageContextInfo.CorrelationId });
            SPStarter._initializeTelemetry(preloadedData);
            sp_core_library_1._SPFlight.initialize(preloadedData.spPageContextInfo.ExpFeatures);
            sp_core_library_1._SPKillSwitch.initialize(preloadedData.spPageContextInfo.killSwitches);
        }
        var spLoader = require('./index');
        spLoader.SPComponentLoader.initialize(new SPSystemJsComponentLoader_1.default());
        spLoader.DeveloperTools.initialize(developerToolsLoader);
        return spLoader.SPComponentLoader.start(preloadedData);
    };
    SPStarter._initializeEnvironment = function () {
        var type;
        if (window.location.hostname === 'localhost' ||
            window['ENVIRONMENTTYPE'] &&
                window['ENVIRONMENTTYPE'] === 'Local') {
            type = sp_core_library_1.EnvironmentType.Local;
        }
        else {
            type = sp_core_library_1.EnvironmentType.SharePoint;
        }
        sp_core_library_1.Environment._initialize({ type: type });
    };
    SPStarter._initializeTelemetry = function (preloadedData) {
        var pageContextInfo = preloadedData.spPageContextInfo;
        sp_telemetry_1._Telemetry.initialize(preloadedData.manifests[0].alias, {
            environment: pageContextInfo.env,
            farmLabel: pageContextInfo.farmLabel,
            clientSideApplicationId: preloadedData.clientSideApplicationId,
            siteSubscriptionId: pageContextInfo.siteSubscriptionId,
            version: !!BUILD_NUMBER ? BUILD_NUMBER : preloadedData.manifests[0].version,
            loginName: pageContextInfo.userLoginName,
            systemUserKey: pageContextInfo.systemUserKey,
            currentUICultureName: pageContextInfo.currentUICultureName,
            correlationId: pageContextInfo.CorrelationId
        });
    };
    return SPStarter;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = SPStarter;
