// @microsoft/sp-build-core-tasks/lib/packageSolution/packageSolution/xmlGenerators/createFeatureXml.js
/**
 * @file createFeatureXml.ts
 * @Copyright (c) Microsoft Corporation.  All rights reserved.
 *
 * Creates an package's Feature.Xml
 */
"use strict";
var XmlGenerator = require("xml");
var lodash_1 = require("lodash");
var createRelationshipsXml_1 = require("./createRelationshipsXml");
var createComponentXml_1 = require("./createComponentXml");
var readSharepointAssets_1 = require("../utils/readSharepointAssets");
var serializationOptions_1 = require("./serializationOptions");
function createFeatureXml(feature, resources) {
    feature = lodash_1.merge({
        version: '1.0.0.0'
    }, feature);
    var featureXmlFile = "feature_" + feature.id + ".xml";
    var featureXml = {
        filename: featureXmlFile,
        /* tslint:disable:no-null-keyword */
        relationships: null,
        /* tslint:enable:no-null-keyword */
        xml: XmlGenerator({
            Feature: [
                {
                    _attr: {
                        xmlns: 'http://schemas.microsoft.com/sharepoint/',
                        Title: feature.title,
                        Description: feature.description,
                        Id: feature.id,
                        Version: feature.version,
                        Scope: 'Web',
                        Hidden: 'FALSE'
                    }
                }
            ]
        }, serializationOptions_1.default)
    };
    var configXml = {
        filename: featureXmlFile + ".config.xml",
        xml: XmlGenerator({
            AppPartConfig: [
                {
                    _attr: {
                        xmlns: 'http://schemas.microsoft.com/sharepoint/2012/app/partconfiguration'
                    }
                },
                {
                    Id: feature.id // @todo needs to be a unique id, different from feature.id
                }
            ]
        }, serializationOptions_1.default)
    };
    var componentsXml = feature.components.map(function (component) {
        return createComponentXml_1.default(component, feature.id);
    });
    // tslint:disable-next-line:max-line-length
    var assets = readSharepointAssets_1.readSharepointAssets(feature.assets, feature.id);
    var relationshipsXml = createRelationshipsXml_1.default(featureXmlFile, [
        {
            type: 'partconfiguration',
            target: configXml.filename
        }
    ].concat(componentsXml.map(function (component) {
        return {
            type: 'feature-elementmanifest',
            target: component.filename
        };
    })).concat((resources.localizedResourcePaths.length > 0) ?
        resources.localizedResourcePaths.map(function (filename) {
            return {
                type: 'content-resource',
                target: filename
            };
        })
        : []).concat((resources.defaultResourcePath)
        ? {
            type: 'content-defaultresource',
            target: resources.defaultResourcePath
        }
        : []).concat(createAssetRelationships(assets)));
    return {
        filename: featureXml.filename,
        xml: featureXml.xml,
        relationships: relationshipsXml,
        config: configXml,
        components: componentsXml,
        assets: assets
    };
}
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = createFeatureXml;
function createAssetRelationships(assets) {
    var relationships = [];
    if (assets) {
        relationships = relationships.concat(createAssetRelationship('feature-elementfile', assets.elementFiles), createAssetRelationship('feature-elementmanifest', assets.elementManifests), createAssetRelationship('feature-upgradeactions', assets.upgradeActions));
    }
    return relationships;
}
function createAssetRelationship(type, files) {
    return files.map(function (file) {
        return {
            type: type,
            target: file.filename
        };
    });
}
