// @microsoft/sp-module-interfaces/lib/manifestSchemas/examples/library_1.manifest.js
"use strict";
var libraryManifest = {
    manifestVersion: 2,
    componentType: 'Library',
    id: '00000000-0000-0000-0000-000000000000',
    alias: 'myLibrary',
    version: '0.0.1',
    loaderConfig: {
        internalModuleBaseUrls: ['https://cdn.net/'],
        entryModuleId: 'main.bundle',
        exportName: 'foobar',
        scriptResources: {
            'frameworkA': {
                type: 'component',
                version: '1.2.3',
                id: '00000000-0000-0000-0000-000000000000'
            },
            'frameworkB': {
                type: 'component',
                version: '1.2.3',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: 'script.js'
            },
            'frameworkC': {
                type: 'component',
                version: 'latest',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: {
                    default: 'script.js'
                }
            },
            'frameworkD': {
                type: 'component',
                version: 'latest',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: {
                    default: 'script.js',
                    debug: 'script.debug.js'
                }
            },
            'frameworkE': {
                type: 'component',
                version: '1.2.3',
                id: '00000000-0000-0000-0000-000000000000',
                shouldNotPreload: true
            },
            'frameworkF': {
                type: 'component',
                version: '1.2.3',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: 'script.js',
                shouldNotPreload: true
            },
            'frameworkG': {
                type: 'component',
                version: 'latest',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: {
                    default: 'script.js'
                },
                shouldNotPreload: true
            },
            'frameworkI': {
                type: 'component',
                version: 'latest',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: {
                    default: 'script.js',
                    debug: 'script.debug.js'
                },
                shouldNotPreload: true
            },
            'frameworkJ': {
                type: 'component',
                version: '1.2.3',
                id: '00000000-0000-0000-0000-000000000000',
                shouldNotPreload: false
            },
            'frameworkK': {
                type: 'component',
                version: '1.2.3',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: 'script.js',
                shouldNotPreload: false
            },
            'frameworkL': {
                type: 'component',
                version: 'latest',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: {
                    default: 'script.js'
                },
                shouldNotPreload: false
            },
            'frameworkM': {
                type: 'component',
                version: 'latest',
                id: '00000000-0000-0000-0000-000000000000',
                failoverPath: {
                    default: 'script.js',
                    debug: 'script.debug.js'
                },
                shouldNotPreload: false
            },
            'main.bundle': {
                type: 'path',
                path: 'bundle.script.js'
            },
            'dependencyA': {
                type: 'path',
                path: {
                    default: 'dependencyA.script.js',
                    debug: 'dependencyA.debug.script.js'
                }
            },
            'dependencyB': {
                type: 'path',
                path: {
                    default: 'dependencyB.script.js'
                }
            },
            'dependencyC': {
                type: 'path',
                path: 'dependencyC.script.js'
            },
            'dependencyD': {
                type: 'path',
                path: {
                    default: 'dependencyD.script.js',
                    debug: 'dependencyD.debug.script.js'
                },
                shouldNotPreload: true
            },
            'dependencyE': {
                type: 'path',
                path: {
                    default: 'dependencyE.script.js'
                },
                shouldNotPreload: true
            },
            'dependencyF': {
                type: 'path',
                path: 'dependencyF.script.js',
                shouldNotPreload: true
            },
            'dependencyG': {
                type: 'path',
                path: {
                    default: 'dependencyG.script.js',
                    debug: 'dependencyG.debug.script.js'
                },
                shouldNotPreload: false
            },
            'dependencyH': {
                type: 'path',
                path: {
                    default: 'dependencyH.script.js'
                },
                shouldNotPreload: false
            },
            'dependencyI': {
                type: 'path',
                path: 'dependencyI.script.js',
                shouldNotPreload: false
            },
            'stringsA': {
                type: 'localizedPath',
                defaultPath: 'stringsA_default.js'
            },
            'stringsB': {
                type: 'localizedPath',
                defaultPath: {
                    default: 'stringsB_default.js',
                    debug: 'stringsB_default.debug.js'
                }
            },
            'stringsC': {
                type: 'localizedPath',
                defaultPath: {
                    default: 'stringsC_default.js'
                }
            },
            'stringsD': {
                type: 'localizedPath',
                defaultPath: 'stringsD_default.js',
                paths: {
                    'en-us': 'stringsD_en-us.js',
                    'fr-fr': 'stringsD_fr-fr.js'
                }
            },
            'stringsE': {
                type: 'localizedPath',
                defaultPath: 'stringsE_default.js',
                paths: {
                    'en-us': {
                        default: 'stringsE_en-us.js'
                    },
                    'fr-fr': 'stringsE_fr-fr.js'
                }
            },
            'stringsF': {
                type: 'localizedPath',
                defaultPath: 'stringsF_default.js',
                paths: {
                    'en-us': {
                        default: 'stringsF_en-us.js',
                        debug: 'stringsF_en-us.debug.js'
                    },
                    'fr-fr': 'stringsF_fr-fr.js'
                }
            },
            'stringsG': {
                type: 'localizedPath',
                defaultPath: 'stringsG_default.js',
                shouldNotPreload: true
            },
            'stringsH': {
                type: 'localizedPath',
                defaultPath: {
                    default: 'stringsH_default.js',
                    debug: 'stringsH_default.debug.js'
                },
                shouldNotPreload: true
            },
            'stringsI': {
                type: 'localizedPath',
                defaultPath: {
                    default: 'stringsI_default.js'
                },
                shouldNotPreload: true
            },
            'stringsJ': {
                type: 'localizedPath',
                defaultPath: 'stringsJ_default.js',
                paths: {
                    'en-us': 'stringsJ_en-us.js',
                    'fr-fr': 'stringsJ_fr-fr.js'
                },
                shouldNotPreload: true
            },
            'stringsK': {
                type: 'localizedPath',
                defaultPath: 'stringsK_default.js',
                paths: {
                    'en-us': {
                        default: 'stringsK_en-us.js'
                    },
                    'fr-fr': 'stringsK_fr-fr.js'
                },
                shouldNotPreload: true
            },
            'stringsL': {
                type: 'localizedPath',
                defaultPath: 'stringsL_default.js',
                paths: {
                    'en-us': {
                        default: 'stringsL_en-us.js',
                        debug: 'stringsL_en-us.debug.js'
                    },
                    'fr-fr': 'stringsL_fr-fr.js'
                },
                shouldNotPreload: true
            },
            'stringsM': {
                type: 'localizedPath',
                defaultPath: 'stringsM_default.js',
                shouldNotPreload: false
            },
            'stringsN': {
                type: 'localizedPath',
                defaultPath: {
                    default: 'stringsN_default.js',
                    debug: 'stringsN_default.debug.js'
                },
                shouldNotPreload: false
            },
            'stringsO': {
                type: 'localizedPath',
                defaultPath: {
                    default: 'stringsO_default.js'
                },
                shouldNotPreload: false
            },
            'stringsP': {
                type: 'localizedPath',
                defaultPath: 'stringsP_default.js',
                paths: {
                    'en-us': 'stringsP_en-us.js',
                    'fr-fr': 'stringsP_fr-fr.js'
                },
                shouldNotPreload: false
            },
            'stringsQ': {
                type: 'localizedPath',
                defaultPath: 'stringsQ_default.js',
                paths: {
                    'en-us': {
                        default: 'stringsQ_en-us.js'
                    },
                    'fr-fr': 'stringsQ_fr-fr.js'
                },
                shouldNotPreload: false
            },
            'stringsR': {
                type: 'localizedPath',
                defaultPath: 'stringsR_default.js',
                paths: {
                    'en-us': {
                        default: 'stringsR_en-us.js',
                        debug: 'stringsR_en-us.debug.js'
                    },
                    'fr-fr': 'stringsR_fr-fr.js'
                },
                shouldNotPreload: false
            },
            'nonAmdA': {
                type: 'path',
                path: 'nonAmdA.js',
                globalName: 'nonAmdA'
            },
            'nonAmdB': {
                type: 'path',
                path: 'nonAmdB.js',
                globalName: 'nonAmdB',
                globalDependencies: ['nonAmdB']
            }
        }
    }
};
module.exports = libraryManifest;
/* tslint:enable:export-name */
